{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e92d84a2",
   "metadata": {},
   "source": [
    "`antman` kernel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "68220ab7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initializing transcriptomics package\n",
      "Initializing HCR package\n",
      "Initializing plotting package\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import Bio.SeqIO as SeqIO\n",
    "import sys\n",
    "from tqdm import tqdm\n",
    "import os \n",
    "import re\n",
    "from pygenomeviz import GenomeViz\n",
    "import scipy.stats as stats\n",
    "import seaborn as sns\n",
    "sys.path.append('../')\n",
    "from transcriptomics import *\n",
    "from hcr import * \n",
    "from plotting import * \n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', 10)\n",
    "white_plotting()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "481cdf2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "tr = load_transcriptome_object(\"../raw-data/OBir/transcriptome/biroi_transcriptome.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "06ee9f2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load or_db\n",
    "or_db = pd.read_excel('../raw-data/OBir/ORs/OR List.xlsx') \n",
    "or_db.loc[or_db.gene_name.isin(['Or5-9E198', 'Or5-9E200', 'Or5-9E201']), 'tandem_array'] = 35"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "546d0259",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>tandem_array</th>\n",
       "      <th>rank</th>\n",
       "      <th>orientation</th>\n",
       "      <th>n_transcripts</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>cds_start</th>\n",
       "      <th>strand</th>\n",
       "      <th>chr</th>\n",
       "      <th>nine_exon</th>\n",
       "      <th>ta_size</th>\n",
       "      <th>ta_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Or5-R2</td>\n",
       "      <td>70</td>\n",
       "      <td>1</td>\n",
       "      <td>3'</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6504395</td>\n",
       "      <td>6510987</td>\n",
       "      <td>6510023.0</td>\n",
       "      <td>-</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Or5-Q1</td>\n",
       "      <td>70</td>\n",
       "      <td>2</td>\n",
       "      <td>5'</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6511625</td>\n",
       "      <td>6514800</td>\n",
       "      <td>6514732.0</td>\n",
       "      <td>-</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>Or5-Orco</td>\n",
       "      <td>72</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10910490</td>\n",
       "      <td>10919026</td>\n",
       "      <td>10911295.0</td>\n",
       "      <td>+</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>Or5-Q2</td>\n",
       "      <td>72</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>10934666</td>\n",
       "      <td>10938750</td>\n",
       "      <td>10934709.0</td>\n",
       "      <td>+</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>Or5-9E1</td>\n",
       "      <td>73</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.0</td>\n",
       "      <td>10980671</td>\n",
       "      <td>10984094</td>\n",
       "      <td>10981538.0</td>\n",
       "      <td>+</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome gene_name  tandem_array  rank orientation  n_transcripts  \\\n",
       "0           1    Or5-R2            70     1         3'             1.0   \n",
       "1           1    Or5-Q1            70     2          5'            1.0   \n",
       "2           1  Or5-Orco            72     3         NaN            1.0   \n",
       "3           1    Or5-Q2            72     4         NaN            2.0   \n",
       "4           1   Or5-9E1            73     5         NaN            7.0   \n",
       "\n",
       "      start       end   cds_start strand   chr nine_exon  ta_size  ta_index  \n",
       "0   6504395   6510987   6510023.0      -  Chr1     False        2         1  \n",
       "1   6511625   6514800   6514732.0      -  Chr1     False        2         2  \n",
       "2  10910490  10919026  10911295.0      +  Chr1     False        2         1  \n",
       "3  10934666  10938750  10934709.0      +  Chr1     False        2         2  \n",
       "4  10980671  10984094  10981538.0      +  Chr1      True        1         1  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# iterate through each row of or_db \n",
    "for i, row in or_db.iterrows(): \n",
    "    gene = tr.get_gene(row.gene_name) \n",
    "    transcript = gene.get_transcript_longest_cds()\n",
    "    bounds = transcript.get_bounds()\n",
    "    or_db.at[i, 'n_transcripts'] = len(gene.transcripts)\n",
    "    or_db.at[i,'start'] = bounds[0]\n",
    "    or_db.at[i,'end'] = bounds[1] \n",
    "    if len(transcript.cds) > 0:\n",
    "        or_db.at[i,'cds_start'] = int(transcript.cds[0].position[0]) if transcript.strand == '+' and len(transcript.cds) > 0 else int(transcript.cds[0].position[1])\n",
    "    or_db.at[i,'strand'] = transcript.strand \n",
    "    or_db.at[i, 'chr'] = transcript.chromosome\n",
    "    or_db.at[i,'nine_exon'] = \"9E\" in row['gene_name']\n",
    "    or_db.at[i,'ta_size'] = or_db.loc[or_db.tandem_array == row['tandem_array']].shape[0]\n",
    "    or_db.at[i,'ta_index'] = or_db.loc[or_db.tandem_array == row['tandem_array']].index.get_loc(i) + 1\n",
    "or_db['start'] = or_db['start'].astype(int)\n",
    "or_db['end'] = or_db['end'].astype(int)\n",
    "or_db['ta_size'] = or_db['ta_size'].astype(int)\n",
    "or_db['ta_index'] = or_db['ta_index'].astype(int)\n",
    "or_db.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "22e93247",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load genome\n",
    "genome_path = \"../raw-data/OBir/genome/Obir.assembly.v5.4.fasta\"\n",
    "genome_seq = SeqIO.to_dict(SeqIO.parse(genome_path, \"fasta\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35bbcbd7",
   "metadata": {},
   "source": [
    "### Load csRNA-seq Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b1928000",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 4 csRNAseq files.\n",
      "['Ooceraea_biroi-WholeAdults_csRNA-r1.negStrand.renamed.bedGraph', 'Ooceraea_biroi-WholeAdults_sRNA-r1.posStrand.renamed.bedGraph', 'Ooceraea_biroi-WholeAdults_sRNA-r1.negStrand.renamed.bedGraph', 'Ooceraea_biroi-WholeAdults_csRNA-r1.posStrand.renamed.bedGraph']\n"
     ]
    }
   ],
   "source": [
    "# Load csRNAseq data\n",
    "csRNAseq_dir = \"../raw-data/OBir/csRNAseq/\"\n",
    "csRNAseq_files = [f for f in os.listdir(csRNAseq_dir) if f.endswith('.bedGraph')]\n",
    "print(f\"Found {len(csRNAseq_files)} csRNAseq files.\")\n",
    "print(csRNAseq_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f6839bd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify adult csRNAseq data, pos and neg strands \n",
    "csWAneg_path = os.path.join(csRNAseq_dir, 'Ooceraea_biroi-WholeAdults_csRNA-r1.negStrand.renamed.bedGraph')\n",
    "csWApos_path = os.path.join(csRNAseq_dir, 'Ooceraea_biroi-WholeAdults_csRNA-r1.posStrand.renamed.bedGraph')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "53e0a518",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>value</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>track type=bedGraph name=\"Ooceraea_biroi-Whole...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>110792.0</td>\n",
       "      <td>110823.0</td>\n",
       "      <td>-1.24</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>122682.0</td>\n",
       "      <td>122683.0</td>\n",
       "      <td>-1.24</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>122683.0</td>\n",
       "      <td>122684.0</td>\n",
       "      <td>-4.96</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>122684.0</td>\n",
       "      <td>122714.0</td>\n",
       "      <td>-6.21</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920935</th>\n",
       "      <td>QOIP01000080.1</td>\n",
       "      <td>13168.0</td>\n",
       "      <td>13172.0</td>\n",
       "      <td>3.72</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920936</th>\n",
       "      <td>QOIP01000080.1</td>\n",
       "      <td>13172.0</td>\n",
       "      <td>13173.0</td>\n",
       "      <td>4.96</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920937</th>\n",
       "      <td>QOIP01000080.1</td>\n",
       "      <td>13173.0</td>\n",
       "      <td>13174.0</td>\n",
       "      <td>8.69</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920938</th>\n",
       "      <td>QOIP01000093.1</td>\n",
       "      <td>6524.0</td>\n",
       "      <td>6549.0</td>\n",
       "      <td>1.24</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920939</th>\n",
       "      <td>QOIP01000093.1</td>\n",
       "      <td>15526.0</td>\n",
       "      <td>15527.0</td>\n",
       "      <td>1.24</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>920940 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    chrom     start       end  \\\n",
       "0       track type=bedGraph name=\"Ooceraea_biroi-Whole...       NaN       NaN   \n",
       "1                                                    Chr1  110792.0  110823.0   \n",
       "2                                                    Chr1  122682.0  122683.0   \n",
       "3                                                    Chr1  122683.0  122684.0   \n",
       "4                                                    Chr1  122684.0  122714.0   \n",
       "...                                                   ...       ...       ...   \n",
       "920935                                     QOIP01000080.1   13168.0   13172.0   \n",
       "920936                                     QOIP01000080.1   13172.0   13173.0   \n",
       "920937                                     QOIP01000080.1   13173.0   13174.0   \n",
       "920938                                     QOIP01000093.1    6524.0    6549.0   \n",
       "920939                                     QOIP01000093.1   15526.0   15527.0   \n",
       "\n",
       "        value strand  \n",
       "0         NaN      -  \n",
       "1       -1.24      -  \n",
       "2       -1.24      -  \n",
       "3       -4.96      -  \n",
       "4       -6.21      -  \n",
       "...       ...    ...  \n",
       "920935   3.72      +  \n",
       "920936   4.96      +  \n",
       "920937   8.69      +  \n",
       "920938   1.24      +  \n",
       "920939   1.24      +  \n",
       "\n",
       "[920940 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csWAneg = pd.read_csv(csWAneg_path, sep=\"\\t\", header=None, names=[\"chrom\", \"start\", \"end\", \"value\"])\n",
    "csWAneg['strand'] = '-' \n",
    "csWApos = pd.read_csv(csWApos_path, sep=\"\\t\", header=None, names=[\"chrom\", \"start\", \"end\", \"value\"])\n",
    "csWApos['strand'] = '+'\n",
    "csWA = pd.concat([csWAneg, csWApos], ignore_index=True)\n",
    "csWA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ae04e0c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>value</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>110792</td>\n",
       "      <td>110823</td>\n",
       "      <td>1.24</td>\n",
       "      <td>-</td>\n",
       "      <td>31</td>\n",
       "      <td>Chr1:110792-110823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>122682</td>\n",
       "      <td>122683</td>\n",
       "      <td>1.24</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>Chr1:122682-122683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>122683</td>\n",
       "      <td>122684</td>\n",
       "      <td>4.96</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>Chr1:122683-122684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>122684</td>\n",
       "      <td>122714</td>\n",
       "      <td>6.21</td>\n",
       "      <td>-</td>\n",
       "      <td>30</td>\n",
       "      <td>Chr1:122684-122714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>138377</td>\n",
       "      <td>138407</td>\n",
       "      <td>2.48</td>\n",
       "      <td>-</td>\n",
       "      <td>30</td>\n",
       "      <td>Chr1:138377-138407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920146</th>\n",
       "      <td>Chr14</td>\n",
       "      <td>8701726</td>\n",
       "      <td>8701732</td>\n",
       "      <td>6.21</td>\n",
       "      <td>+</td>\n",
       "      <td>6</td>\n",
       "      <td>Chr14:8701726-8701732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920147</th>\n",
       "      <td>Chr14</td>\n",
       "      <td>8701732</td>\n",
       "      <td>8701733</td>\n",
       "      <td>4.97</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>Chr14:8701732-8701733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920148</th>\n",
       "      <td>Chr14</td>\n",
       "      <td>8701733</td>\n",
       "      <td>8701756</td>\n",
       "      <td>3.73</td>\n",
       "      <td>+</td>\n",
       "      <td>23</td>\n",
       "      <td>Chr14:8701733-8701756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920149</th>\n",
       "      <td>Chr14</td>\n",
       "      <td>8702324</td>\n",
       "      <td>8702344</td>\n",
       "      <td>1.25</td>\n",
       "      <td>+</td>\n",
       "      <td>20</td>\n",
       "      <td>Chr14:8702324-8702344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920150</th>\n",
       "      <td>Chr14</td>\n",
       "      <td>8719033</td>\n",
       "      <td>8719034</td>\n",
       "      <td>1.25</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>Chr14:8719033-8719034</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>920151 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom    start      end  value strand  length               location\n",
       "0        Chr1   110792   110823   1.24      -      31     Chr1:110792-110823\n",
       "1        Chr1   122682   122683   1.24      -       1     Chr1:122682-122683\n",
       "2        Chr1   122683   122684   4.96      -       1     Chr1:122683-122684\n",
       "3        Chr1   122684   122714   6.21      -      30     Chr1:122684-122714\n",
       "4        Chr1   138377   138407   2.48      -      30     Chr1:138377-138407\n",
       "...       ...      ...      ...    ...    ...     ...                    ...\n",
       "920146  Chr14  8701726  8701732   6.21      +       6  Chr14:8701726-8701732\n",
       "920147  Chr14  8701732  8701733   4.97      +       1  Chr14:8701732-8701733\n",
       "920148  Chr14  8701733  8701756   3.73      +      23  Chr14:8701733-8701756\n",
       "920149  Chr14  8702324  8702344   1.25      +      20  Chr14:8702324-8702344\n",
       "920150  Chr14  8719033  8719034   1.25      +       1  Chr14:8719033-8719034\n",
       "\n",
       "[920151 rows x 7 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop rows with NaN in value column, reset index, also drop rows without \"Chr\" in the chrom \n",
    "csWA = csWA.dropna(subset=['value'])\n",
    "csWA = csWA[csWA['chrom'].str.contains(\"Chr\")]\n",
    "csWA = csWA.reset_index(drop=True)  \n",
    "csWA['start'] = csWA['start'].astype(int)\n",
    "csWA['end'] = csWA['end'].astype(int)\n",
    "csWA['value'] = np.abs(csWA['value'])  # Ensure values are positive\n",
    "csWA['length'] = csWA['end'] - csWA['start']\n",
    "csWA['location'] = csWA['chrom'] + ':' + csWA['start'].astype(str) + '-' + csWA['end'].astype(str)\n",
    "csWA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a7c0e59",
   "metadata": {},
   "source": [
    "### Specify windows upstream of ORs to search for csRNA-seq reads"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a4d0390",
   "metadata": {},
   "source": [
    "Take a 1kbp window upstream of the CDS start site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "726b4807",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>chrom_num</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>6510023</td>\n",
       "      <td>6511023</td>\n",
       "      <td>-</td>\n",
       "      <td>Or5-R2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>6514732</td>\n",
       "      <td>6515732</td>\n",
       "      <td>-</td>\n",
       "      <td>Or5-Q1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>10910295</td>\n",
       "      <td>10911295</td>\n",
       "      <td>+</td>\n",
       "      <td>Or5-Orco</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>10933709</td>\n",
       "      <td>10934709</td>\n",
       "      <td>+</td>\n",
       "      <td>Or5-Q2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>10980538</td>\n",
       "      <td>10981538</td>\n",
       "      <td>+</td>\n",
       "      <td>Or5-9E1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>Chr12</td>\n",
       "      <td>12</td>\n",
       "      <td>8807087</td>\n",
       "      <td>8808087</td>\n",
       "      <td>+</td>\n",
       "      <td>Or5-L35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>494</th>\n",
       "      <td>Chr12</td>\n",
       "      <td>12</td>\n",
       "      <td>12304782</td>\n",
       "      <td>12305782</td>\n",
       "      <td>+</td>\n",
       "      <td>Or5-9E350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>Chr12</td>\n",
       "      <td>12</td>\n",
       "      <td>13237580</td>\n",
       "      <td>13238580</td>\n",
       "      <td>-</td>\n",
       "      <td>Or5-9E351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>Chr12</td>\n",
       "      <td>12</td>\n",
       "      <td>13619576</td>\n",
       "      <td>13620576</td>\n",
       "      <td>-</td>\n",
       "      <td>Or5-9Efrag3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>Chr14</td>\n",
       "      <td>14</td>\n",
       "      <td>1247228</td>\n",
       "      <td>1248228</td>\n",
       "      <td>+</td>\n",
       "      <td>Or5-9E352</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>498 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom  chrom_num     start       end strand    gene_name\n",
       "0     Chr1          1   6510023   6511023      -       Or5-R2\n",
       "1     Chr1          1   6514732   6515732      -       Or5-Q1\n",
       "2     Chr1          1  10910295  10911295      +     Or5-Orco\n",
       "3     Chr1          1  10933709  10934709      +       Or5-Q2\n",
       "4     Chr1          1  10980538  10981538      +      Or5-9E1\n",
       "..     ...        ...       ...       ...    ...          ...\n",
       "493  Chr12         12   8807087   8808087      +      Or5-L35\n",
       "494  Chr12         12  12304782  12305782      +    Or5-9E350\n",
       "495  Chr12         12  13237580  13238580      -    Or5-9E351\n",
       "496  Chr12         12  13619576  13620576      -  Or5-9Efrag3\n",
       "497  Chr14         14   1247228   1248228      +    Or5-9E352\n",
       "\n",
       "[498 rows x 6 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Specify regions of interest (1kbp upstream of each OR's CDS start)\n",
    "dist_upstream = 1000 \n",
    "\n",
    "OR_rois = [] \n",
    "\n",
    "for i, row in or_db.iterrows():\n",
    "    if pd.isna(row['chr']) or pd.isna(row['cds_start']):\n",
    "        continue\n",
    "    chrom = row['chr']\n",
    "    chrom_num = int(chrom.replace('Chr', ''))  # Convert 'Chr1' to 1, 'Chr2' to 2, etc.\n",
    "    strand = row['strand']\n",
    "    start = row['cds_start'] - dist_upstream if strand == '+' else row['cds_start']\n",
    "    end = row['cds_start'] if strand == '+' else row['cds_start'] + dist_upstream\n",
    "    roi = {'chrom': chrom, 'chrom_num': chrom_num, 'start': start, 'end': end, 'strand': strand, 'gene_name': row['gene_name']}\n",
    "    OR_rois.append(roi)\n",
    "\n",
    "OR_rois = pd.DataFrame(OR_rois)\n",
    "# Convert start and end to ints\n",
    "OR_rois['start'] = OR_rois['start'].astype(int)\n",
    "OR_rois['end'] = OR_rois['end'].astype(int)\n",
    "OR_rois"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "20103e85",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert OR_rois to a numpy array \n",
    "OR_rois_array = OR_rois[['chrom', 'start', 'end']].to_numpy()\n",
    "\n",
    "# Use the first column to filter \n",
    "OR_rois_array_dict = {}\n",
    "for chrom, start, end in OR_rois_array:\n",
    "    if chrom not in OR_rois_array_dict:\n",
    "        OR_rois_array_dict[chrom] = []\n",
    "    OR_rois_array_dict[chrom].append((start, end))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9b77abe8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['Chr1', 'Chr2', 'Chr3', 'Chr4', 'Chr5', 'Chr6', 'Chr7', 'Chr8', 'Chr9', 'Chr10', 'Chr11', 'Chr12', 'Chr14'])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "OR_rois_array_dict.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dce520f",
   "metadata": {},
   "source": [
    "### Take intersection to remove non-OR-related csRNA-seq Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bfc40453",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop rows where chrom is not in OR_rois_array_dict keys \n",
    "csWA = csWA[csWA['chrom'].isin(OR_rois_array_dict.keys())].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cc8eb675",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing csRNAseq data: 100%|██████████| 893560/893560 [00:42<00:00, 21101.50it/s]\n"
     ]
    }
   ],
   "source": [
    "# Take intersection of csRNAseq data with OR regions of interest\n",
    "from tqdm import tqdm\n",
    "\n",
    "# Initialize a list to store overlapping regions\n",
    "overlapping = np.zeros(csWA.shape[0], dtype=bool)\n",
    "\n",
    "for i, row in tqdm(csWA.iterrows(), total=csWA.shape[0], desc=\"Processing csRNAseq data\"):\n",
    "    OR_rois_chrom = OR_rois_array_dict[row['chrom']]\n",
    "    for start, end in OR_rois_chrom:\n",
    "        if start <= row['start'] and row['end'] <= end:\n",
    "            # Overlapping region found\n",
    "            overlapping[i] = True\n",
    "            break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fbbb9dfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of overlapping regions: 869 out of 893560 total regions.\n"
     ]
    }
   ],
   "source": [
    "# Look for overlapping regions\n",
    "csWA['overlapping'] = overlapping\n",
    "print(f\"Number of overlapping regions: {np.sum(overlapping)} out of {csWA.shape[0]} total regions.\")\n",
    "\n",
    "# Filter csWA to keep only overlapping regions\n",
    "csWA = csWA[csWA['overlapping']].reset_index(drop=True)\n",
    "\n",
    "# Drop the 'overlapping' column\n",
    "csWA = csWA.drop(columns=['overlapping'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3452194c",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_1 = csWA.copy() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6107bfd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "csWA = save_1.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d9d629bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing csRNAseq data: 100%|██████████| 869/869 [00:02<00:00, 331.37it/s]\n"
     ]
    }
   ],
   "source": [
    "# Save gene name and strand information from OR_rois to csWA\n",
    "for i, row in tqdm(csWA.iterrows(), total=csWA.shape[0], desc=\"Processing csRNAseq data\"):\n",
    "    # Find row in OR_rois \n",
    "    for j, or_row in OR_rois.iterrows():\n",
    "        if row['chrom'] == or_row['chrom'] and row['start'] >= or_row['start'] and row['end'] <= or_row['end']:\n",
    "            # Found a matching OR region\n",
    "            csWA.at[i, 'gene_name'] = or_row['gene_name']\n",
    "            csWA.at[i, 'gene_strand'] = or_row['strand']\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4a1cad5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>value</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "      <th>location</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrom, start, end, value, strand, length, location, gene_name, gene_strand]\n",
       "Index: []"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Look for Nan Rows \n",
    "csWA[csWA.isna().any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d820d93a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>value</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "      <th>location</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>16155170</td>\n",
       "      <td>16155188</td>\n",
       "      <td>1.23</td>\n",
       "      <td>-</td>\n",
       "      <td>18</td>\n",
       "      <td>Chr1:16155170-16155188</td>\n",
       "      <td>Or5-D2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>16155188</td>\n",
       "      <td>16155212</td>\n",
       "      <td>2.47</td>\n",
       "      <td>-</td>\n",
       "      <td>24</td>\n",
       "      <td>Chr1:16155188-16155212</td>\n",
       "      <td>Or5-D2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>16155212</td>\n",
       "      <td>16155214</td>\n",
       "      <td>1.23</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>Chr1:16155212-16155214</td>\n",
       "      <td>Or5-D2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>16155214</td>\n",
       "      <td>16155224</td>\n",
       "      <td>2.47</td>\n",
       "      <td>-</td>\n",
       "      <td>10</td>\n",
       "      <td>Chr1:16155214-16155224</td>\n",
       "      <td>Or5-D2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>16155224</td>\n",
       "      <td>16155252</td>\n",
       "      <td>1.23</td>\n",
       "      <td>-</td>\n",
       "      <td>28</td>\n",
       "      <td>Chr1:16155224-16155252</td>\n",
       "      <td>Or5-D2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>864</th>\n",
       "      <td>Chr12</td>\n",
       "      <td>8808077</td>\n",
       "      <td>8808079</td>\n",
       "      <td>73.23</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>Chr12:8808077-8808079</td>\n",
       "      <td>Or5-L35</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865</th>\n",
       "      <td>Chr12</td>\n",
       "      <td>8808079</td>\n",
       "      <td>8808080</td>\n",
       "      <td>71.99</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>Chr12:8808079-8808080</td>\n",
       "      <td>Or5-L35</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>866</th>\n",
       "      <td>Chr12</td>\n",
       "      <td>8808080</td>\n",
       "      <td>8808082</td>\n",
       "      <td>70.75</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>Chr12:8808080-8808082</td>\n",
       "      <td>Or5-L35</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>867</th>\n",
       "      <td>Chr12</td>\n",
       "      <td>8808082</td>\n",
       "      <td>8808084</td>\n",
       "      <td>68.27</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>Chr12:8808082-8808084</td>\n",
       "      <td>Or5-L35</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>868</th>\n",
       "      <td>Chr12</td>\n",
       "      <td>12305729</td>\n",
       "      <td>12305767</td>\n",
       "      <td>1.24</td>\n",
       "      <td>+</td>\n",
       "      <td>38</td>\n",
       "      <td>Chr12:12305729-12305767</td>\n",
       "      <td>Or5-9E350</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>869 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end  value strand  length                 location  \\\n",
       "0     Chr1  16155170  16155188   1.23      -      18   Chr1:16155170-16155188   \n",
       "1     Chr1  16155188  16155212   2.47      -      24   Chr1:16155188-16155212   \n",
       "2     Chr1  16155212  16155214   1.23      -       2   Chr1:16155212-16155214   \n",
       "3     Chr1  16155214  16155224   2.47      -      10   Chr1:16155214-16155224   \n",
       "4     Chr1  16155224  16155252   1.23      -      28   Chr1:16155224-16155252   \n",
       "..     ...       ...       ...    ...    ...     ...                      ...   \n",
       "864  Chr12   8808077   8808079  73.23      +       2    Chr12:8808077-8808079   \n",
       "865  Chr12   8808079   8808080  71.99      +       1    Chr12:8808079-8808080   \n",
       "866  Chr12   8808080   8808082  70.75      +       2    Chr12:8808080-8808082   \n",
       "867  Chr12   8808082   8808084  68.27      +       2    Chr12:8808082-8808084   \n",
       "868  Chr12  12305729  12305767   1.24      +      38  Chr12:12305729-12305767   \n",
       "\n",
       "     gene_name gene_strand  \n",
       "0       Or5-D2           +  \n",
       "1       Or5-D2           +  \n",
       "2       Or5-D2           +  \n",
       "3       Or5-D2           +  \n",
       "4       Or5-D2           +  \n",
       "..         ...         ...  \n",
       "864    Or5-L35           +  \n",
       "865    Or5-L35           +  \n",
       "866    Or5-L35           +  \n",
       "867    Or5-L35           +  \n",
       "868  Or5-9E350           +  \n",
       "\n",
       "[869 rows x 9 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csWA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbbb5726",
   "metadata": {},
   "source": [
    "### Merge Overlapping Reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "acef028c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>value</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "      <th>location</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>Chr9</td>\n",
       "      <td>14870015</td>\n",
       "      <td>14870058</td>\n",
       "      <td>1.23</td>\n",
       "      <td>-</td>\n",
       "      <td>43</td>\n",
       "      <td>Chr9:14870015-14870058</td>\n",
       "      <td>Or5-E11</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>Chr9</td>\n",
       "      <td>14870058</td>\n",
       "      <td>14870063</td>\n",
       "      <td>2.48</td>\n",
       "      <td>-</td>\n",
       "      <td>5</td>\n",
       "      <td>Chr9:14870058-14870063</td>\n",
       "      <td>Or5-E11</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>Chr9</td>\n",
       "      <td>14870063</td>\n",
       "      <td>14870066</td>\n",
       "      <td>8.68</td>\n",
       "      <td>-</td>\n",
       "      <td>3</td>\n",
       "      <td>Chr9:14870063-14870066</td>\n",
       "      <td>Or5-E11</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>Chr9</td>\n",
       "      <td>14870066</td>\n",
       "      <td>14870080</td>\n",
       "      <td>9.92</td>\n",
       "      <td>-</td>\n",
       "      <td>14</td>\n",
       "      <td>Chr9:14870066-14870080</td>\n",
       "      <td>Or5-E11</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>Chr9</td>\n",
       "      <td>14870080</td>\n",
       "      <td>14870103</td>\n",
       "      <td>11.16</td>\n",
       "      <td>-</td>\n",
       "      <td>23</td>\n",
       "      <td>Chr9:14870080-14870103</td>\n",
       "      <td>Or5-E11</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>Chr9</td>\n",
       "      <td>14870103</td>\n",
       "      <td>14870127</td>\n",
       "      <td>1.23</td>\n",
       "      <td>-</td>\n",
       "      <td>24</td>\n",
       "      <td>Chr9:14870103-14870127</td>\n",
       "      <td>Or5-E11</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>Chr9</td>\n",
       "      <td>14870131</td>\n",
       "      <td>14870165</td>\n",
       "      <td>1.23</td>\n",
       "      <td>-</td>\n",
       "      <td>34</td>\n",
       "      <td>Chr9:14870131-14870165</td>\n",
       "      <td>Or5-E11</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom     start       end  value strand  length                location  \\\n",
       "146  Chr9  14870015  14870058   1.23      -      43  Chr9:14870015-14870058   \n",
       "147  Chr9  14870058  14870063   2.48      -       5  Chr9:14870058-14870063   \n",
       "148  Chr9  14870063  14870066   8.68      -       3  Chr9:14870063-14870066   \n",
       "149  Chr9  14870066  14870080   9.92      -      14  Chr9:14870066-14870080   \n",
       "150  Chr9  14870080  14870103  11.16      -      23  Chr9:14870080-14870103   \n",
       "151  Chr9  14870103  14870127   1.23      -      24  Chr9:14870103-14870127   \n",
       "152  Chr9  14870131  14870165   1.23      -      34  Chr9:14870131-14870165   \n",
       "\n",
       "    gene_name gene_strand  \n",
       "146   Or5-E11           -  \n",
       "147   Or5-E11           -  \n",
       "148   Or5-E11           -  \n",
       "149   Or5-E11           -  \n",
       "150   Or5-E11           -  \n",
       "151   Or5-E11           -  \n",
       "152   Or5-E11           -  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find where gene_name is Or5-E11\n",
    "csWA[csWA['gene_name'] == 'Or5-E11']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c21cc4a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>value</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "      <th>location</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>10910413</td>\n",
       "      <td>10910497</td>\n",
       "      <td>17.567543</td>\n",
       "      <td>+</td>\n",
       "      <td>84</td>\n",
       "      <td>Chr1:10910413-10910497</td>\n",
       "      <td>Or5-Orco</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>10934665</td>\n",
       "      <td>10934691</td>\n",
       "      <td>2.480000</td>\n",
       "      <td>+</td>\n",
       "      <td>26</td>\n",
       "      <td>Chr1:10934665-10934691</td>\n",
       "      <td>Or5-Q2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>16155170</td>\n",
       "      <td>16155252</td>\n",
       "      <td>1.645712</td>\n",
       "      <td>-</td>\n",
       "      <td>82</td>\n",
       "      <td>Chr1:16155170-16155252</td>\n",
       "      <td>Or5-D2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>16155761</td>\n",
       "      <td>16155791</td>\n",
       "      <td>2.470000</td>\n",
       "      <td>-</td>\n",
       "      <td>30</td>\n",
       "      <td>Chr1:16155761-16155791</td>\n",
       "      <td>Or5-D2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>16155928</td>\n",
       "      <td>16155955</td>\n",
       "      <td>1.240000</td>\n",
       "      <td>+</td>\n",
       "      <td>27</td>\n",
       "      <td>Chr1:16155928-16155955</td>\n",
       "      <td>Or5-D2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>Chr9</td>\n",
       "      <td>14882134</td>\n",
       "      <td>14882154</td>\n",
       "      <td>3.720000</td>\n",
       "      <td>-</td>\n",
       "      <td>20</td>\n",
       "      <td>Chr9:14882134-14882154</td>\n",
       "      <td>Or5-E14</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>Chr9</td>\n",
       "      <td>14888724</td>\n",
       "      <td>14888766</td>\n",
       "      <td>2.480000</td>\n",
       "      <td>-</td>\n",
       "      <td>42</td>\n",
       "      <td>Chr9:14888724-14888766</td>\n",
       "      <td>Or5-E15</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>Chr9</td>\n",
       "      <td>14888785</td>\n",
       "      <td>14888836</td>\n",
       "      <td>56.601292</td>\n",
       "      <td>-</td>\n",
       "      <td>51</td>\n",
       "      <td>Chr9:14888785-14888836</td>\n",
       "      <td>Or5-E15</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>Chr9</td>\n",
       "      <td>14888845</td>\n",
       "      <td>14888887</td>\n",
       "      <td>2.428719</td>\n",
       "      <td>-</td>\n",
       "      <td>42</td>\n",
       "      <td>Chr9:14888845-14888887</td>\n",
       "      <td>Or5-E15</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>Chr9</td>\n",
       "      <td>14901941</td>\n",
       "      <td>14901968</td>\n",
       "      <td>4.304906</td>\n",
       "      <td>-</td>\n",
       "      <td>27</td>\n",
       "      <td>Chr9:14901941-14901968</td>\n",
       "      <td>Or5-E17</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>289 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom     start       end      value strand  length  \\\n",
       "0    Chr1  10910413  10910497  17.567543      +      84   \n",
       "1    Chr1  10934665  10934691   2.480000      +      26   \n",
       "2    Chr1  16155170  16155252   1.645712      -      82   \n",
       "3    Chr1  16155761  16155791   2.470000      -      30   \n",
       "4    Chr1  16155928  16155955   1.240000      +      27   \n",
       "..    ...       ...       ...        ...    ...     ...   \n",
       "284  Chr9  14882134  14882154   3.720000      -      20   \n",
       "285  Chr9  14888724  14888766   2.480000      -      42   \n",
       "286  Chr9  14888785  14888836  56.601292      -      51   \n",
       "287  Chr9  14888845  14888887   2.428719      -      42   \n",
       "288  Chr9  14901941  14901968   4.304906      -      27   \n",
       "\n",
       "                   location gene_name gene_strand  \n",
       "0    Chr1:10910413-10910497  Or5-Orco           +  \n",
       "1    Chr1:10934665-10934691    Or5-Q2           +  \n",
       "2    Chr1:16155170-16155252    Or5-D2           +  \n",
       "3    Chr1:16155761-16155791    Or5-D2           +  \n",
       "4    Chr1:16155928-16155955    Or5-D2           +  \n",
       "..                      ...       ...         ...  \n",
       "284  Chr9:14882134-14882154   Or5-E14           -  \n",
       "285  Chr9:14888724-14888766   Or5-E15           -  \n",
       "286  Chr9:14888785-14888836   Or5-E15           -  \n",
       "287  Chr9:14888845-14888887   Or5-E15           -  \n",
       "288  Chr9:14901941-14901968   Or5-E17           -  \n",
       "\n",
       "[289 rows x 9 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assuming csWA is already sorted by chrom and start\n",
    "csWA = csWA.sort_values(by=['chrom', 'start']).reset_index(drop=True)\n",
    "\n",
    "# Initialize a list to store the merged rows\n",
    "merged_rows = []\n",
    "\n",
    "# Initialize the previous row as None to start the first merge\n",
    "previous_row = None\n",
    "\n",
    "# Iterate through each row\n",
    "for i in range(len(csWA)):\n",
    "    row = csWA.iloc[i].copy()\n",
    "    \n",
    "    if previous_row is None:\n",
    "        # The first region becomes the initial \"merged\" row\n",
    "        previous_row = row.copy() \n",
    "        continue\n",
    "    \n",
    "    # Check if the current row overlaps with the previous row (current merged region)\n",
    "    if (row['chrom'] == previous_row['chrom'] and\n",
    "        row['strand'] == previous_row['strand'] and\n",
    "        row['start'] <= previous_row['end']):\n",
    "        \n",
    "        # Merge: Extend the end of the merged region\n",
    "        previous_row['end'] = max(previous_row['end'], row['end'])\n",
    "        previous_row['length'] = previous_row['end'] - previous_row['start']\n",
    "        previous_row['location'] = previous_row['chrom'] + ':' + str(previous_row['start']) + '-' + str(previous_row['end'])\n",
    "\n",
    "        # Take the weighted mean of 'value'\n",
    "        previous_row['value'] = (previous_row['value'] * previous_row['length'] + row['value'] * row['length']) / (previous_row['length'] + row['length'])\n",
    "    else:\n",
    "        # If no overlap, add the previous row to the merged rows and start a new merged region\n",
    "        merged_rows.append(previous_row)\n",
    "        previous_row = row.copy()  # Update the previous_row to the current row\n",
    "\n",
    "# After the loop, append the last region (since it hasn't been added yet)\n",
    "merged_rows.append(previous_row)\n",
    "\n",
    "# Convert the list of merged rows back into a DataFrame\n",
    "csWA_merged = pd.DataFrame(merged_rows).reset_index(drop=True)\n",
    "csWA_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1d6da75f",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_2 = csWA_merged.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "9aa20739",
   "metadata": {},
   "outputs": [],
   "source": [
    "csWA = csWA_merged.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "dbcb6233",
   "metadata": {},
   "outputs": [],
   "source": [
    "csWA = csWA.sort_values(by=['value'], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b4aba32e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>value</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "      <th>location</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>Chr4</td>\n",
       "      <td>10527180</td>\n",
       "      <td>10527236</td>\n",
       "      <td>562.590346</td>\n",
       "      <td>+</td>\n",
       "      <td>56</td>\n",
       "      <td>Chr4:10527180-10527236</td>\n",
       "      <td>Or5-9E108</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Chr10</td>\n",
       "      <td>1963639</td>\n",
       "      <td>1963688</td>\n",
       "      <td>473.125627</td>\n",
       "      <td>+</td>\n",
       "      <td>49</td>\n",
       "      <td>Chr10:1963639-1963688</td>\n",
       "      <td>Or5-9E241</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Chr12</td>\n",
       "      <td>8761335</td>\n",
       "      <td>8761369</td>\n",
       "      <td>311.071569</td>\n",
       "      <td>+</td>\n",
       "      <td>34</td>\n",
       "      <td>Chr12:8761335-8761369</td>\n",
       "      <td>Or5-L22</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>Chr4</td>\n",
       "      <td>10603206</td>\n",
       "      <td>10603259</td>\n",
       "      <td>101.522030</td>\n",
       "      <td>+</td>\n",
       "      <td>53</td>\n",
       "      <td>Chr4:10603206-10603259</td>\n",
       "      <td>Or5-9E118</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>Chr7</td>\n",
       "      <td>5138097</td>\n",
       "      <td>5138138</td>\n",
       "      <td>95.484292</td>\n",
       "      <td>-</td>\n",
       "      <td>41</td>\n",
       "      <td>Chr7:5138097-5138138</td>\n",
       "      <td>Or5-U64</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>Chr4</td>\n",
       "      <td>10543210</td>\n",
       "      <td>10543252</td>\n",
       "      <td>64.611289</td>\n",
       "      <td>+</td>\n",
       "      <td>42</td>\n",
       "      <td>Chr4:10543210-10543252</td>\n",
       "      <td>Or5-9E110</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Chr10</td>\n",
       "      <td>1976532</td>\n",
       "      <td>1976626</td>\n",
       "      <td>62.659864</td>\n",
       "      <td>+</td>\n",
       "      <td>94</td>\n",
       "      <td>Chr10:1976532-1976626</td>\n",
       "      <td>Or5-9E243</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>Chr4</td>\n",
       "      <td>10566339</td>\n",
       "      <td>10566381</td>\n",
       "      <td>60.505970</td>\n",
       "      <td>+</td>\n",
       "      <td>42</td>\n",
       "      <td>Chr4:10566339-10566381</td>\n",
       "      <td>Or5-9E113</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>Chr4</td>\n",
       "      <td>10576216</td>\n",
       "      <td>10576262</td>\n",
       "      <td>59.371470</td>\n",
       "      <td>+</td>\n",
       "      <td>46</td>\n",
       "      <td>Chr4:10576216-10576262</td>\n",
       "      <td>Or5-9E114</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>Chr12</td>\n",
       "      <td>8808055</td>\n",
       "      <td>8808084</td>\n",
       "      <td>57.083637</td>\n",
       "      <td>+</td>\n",
       "      <td>29</td>\n",
       "      <td>Chr12:8808055-8808084</td>\n",
       "      <td>Or5-L35</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end       value strand  length  \\\n",
       "184   Chr4  10527180  10527236  562.590346      +      56   \n",
       "20   Chr10   1963639   1963688  473.125627      +      49   \n",
       "109  Chr12   8761335   8761369  311.071569      +      34   \n",
       "199   Chr4  10603206  10603259  101.522030      +      53   \n",
       "264   Chr7   5138097   5138138   95.484292      -      41   \n",
       "186   Chr4  10543210  10543252   64.611289      +      42   \n",
       "21   Chr10   1976532   1976626   62.659864      +      94   \n",
       "189   Chr4  10566339  10566381   60.505970      +      42   \n",
       "191   Chr4  10576216  10576262   59.371470      +      46   \n",
       "114  Chr12   8808055   8808084   57.083637      +      29   \n",
       "\n",
       "                   location  gene_name gene_strand  \n",
       "184  Chr4:10527180-10527236  Or5-9E108           -  \n",
       "20    Chr10:1963639-1963688  Or5-9E241           -  \n",
       "109   Chr12:8761335-8761369    Or5-L22           +  \n",
       "199  Chr4:10603206-10603259  Or5-9E118           -  \n",
       "264    Chr7:5138097-5138138    Or5-U64           +  \n",
       "186  Chr4:10543210-10543252  Or5-9E110           -  \n",
       "21    Chr10:1976532-1976626  Or5-9E243           -  \n",
       "189  Chr4:10566339-10566381  Or5-9E113           -  \n",
       "191  Chr4:10576216-10576262  Or5-9E114           -  \n",
       "114   Chr12:8808055-8808084    Or5-L35           +  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csWA.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "7fa51511",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>value</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "      <th>location</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>Chr9</td>\n",
       "      <td>14870015</td>\n",
       "      <td>14870127</td>\n",
       "      <td>4.212079</td>\n",
       "      <td>-</td>\n",
       "      <td>112</td>\n",
       "      <td>Chr9:14870015-14870127</td>\n",
       "      <td>Or5-E11</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>Chr9</td>\n",
       "      <td>14870131</td>\n",
       "      <td>14870165</td>\n",
       "      <td>1.230000</td>\n",
       "      <td>-</td>\n",
       "      <td>34</td>\n",
       "      <td>Chr9:14870131-14870165</td>\n",
       "      <td>Or5-E11</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom     start       end     value strand  length  \\\n",
       "276  Chr9  14870015  14870127  4.212079      -     112   \n",
       "277  Chr9  14870131  14870165  1.230000      -      34   \n",
       "\n",
       "                   location gene_name gene_strand  \n",
       "276  Chr9:14870015-14870127   Or5-E11           -  \n",
       "277  Chr9:14870131-14870165   Or5-E11           -  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find where gene_name is Or5-E11\n",
    "csWA[csWA['gene_name'] == 'Or5-E11']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "10864748",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>value</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "      <th>location</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>Chr2</td>\n",
       "      <td>3732972</td>\n",
       "      <td>3732997</td>\n",
       "      <td>1.733369</td>\n",
       "      <td>+</td>\n",
       "      <td>25</td>\n",
       "      <td>Chr2:3732972-3732997</td>\n",
       "      <td>Or5-Y1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom    start      end     value strand  length              location  \\\n",
       "121  Chr2  3732972  3732997  1.733369      +      25  Chr2:3732972-3732997   \n",
       "\n",
       "    gene_name gene_strand  \n",
       "121    Or5-Y1           +  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find where gene_name is Or5-Y1\n",
    "csWA[csWA['gene_name'] == 'Or5-Y1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3da532e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get strand relative to gene \n",
    "csWA['rel_strand'] = np.where(csWA['strand'] == csWA['gene_strand'], 'sense', 'antisense')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "61924be1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get sequences from genome\n",
    "csWA['sequence'] = csWA.apply(lambda row: get_sequence(genome_seq, row['chrom'], row['start'], row['end'], row['strand']), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "fbd1d796",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop rows with length < 9\n",
    "csWA = csWA[csWA['length'] >= 9].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e898cd94",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'Frequency')"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA50AAAHUCAYAAAC55EzHAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8fJSN1AAAACXBIWXMAAA9hAAAPYQGoP6dpAABJz0lEQVR4nO3deXQUZdr+8SuQpLMnEDCACWETJMALyja4sG8ieFhFJEAIjoAIsowKLkAYIeqoLIPoK7KJLCqIGtkGgYiyIzOoYRSMQiL7lgUCgSTP7w9/9EubBMJSNFS+n3PqHPqp6ur76TR95+qqrngYY4wAAAAAALBACXcXAAAAAACwL0InAAAAAMAyhE4AAAAAgGUInQAAAAAAyxA6AQAAAACWIXQCAAAAACxD6AQAAAAAWIbQCQAAAACwDKETAAAAAGAZQicAAAAAwDKETgAAAACAZQidAAAAAADLEDoBAAAAAJYhdAIAAAAALEPoBAAAAABYhtAJAAAAALAMoRMAAAAAYBlCJwAAAADAMoROAAAAAIBlCJ0AAAAAAMsQOgEAAAAAliF0AgAAAAAsQ+gEAAAAAFiG0AkAAAAAsAyhEwAAAABgGUInAAAAAMAyhE4AAAAAgGUInQAAAAAAyxA6AQAAAACWIXQCAAAAACxD6AQAAAAAWIbQCQAAAACwDKETAAAAAGAZQicAAAAAwDKETgAAAACAZQidAAAAAADLEDoBAAAAAJYhdAIAAAAALEPoBAAAAABYhtAJAAAAALAMoRMAAAAAYBlCJwAAAADAMoROAAAAAIBlCJ0AAAAAAMsQOgEAAAAAliF0AgAAAAAsQ+gEAAAAAFiG0AkAAAAAsAyhE7hKxhh3lwAAKMboQ3AXXnu4VoRO2EpMTIwqVapU6PrmzZurefPmhd6+ko0bN6pjx47XXqBN/P7772rWrJl8fHx0xx13KCsr64bu3xijsLAwde7cOd+6KVOmyMPDQy1atMi3bvr06fLw8NBPP/3kMt60aVN5eHjoo48+uuzjZmVlKTg4WA8//HCh2xw/flze3t56/vnnizQXDw8PjR8/vkjbArj90YduDrv2oYtuxd4xceJEvfHGG87b48ePl4eHhxsrwu2E0IlibcaMGZoxY0aRt585c6aSkpIsrOj2MGXKFG3atEkffPCBli1bJj8/vxu6fw8PD7Vs2VIbN27Mt27VqlUKDQ3Vxo0blZmZ6bJuw4YNioiI0N133+0c++WXX/TNN9+oTp06evfddy/7uH5+furVq5f+9a9/6dixYwVus3DhQl24cEGxsbHXMDMAcEUfujZ27UO3spdeeklnzpxxdxm4TRE6UaxFRUUpKirK3WXcdk6cOKEKFSro0Ucf1f3332/JY7Ru3VrHjx/Xnj17nGPnzp3Thg0b9Oyzz+rChQtat26dy32++eYbtW3b1mVs9uzZCg8P18svv6zExMR8nz7/WWxsrHJycgr9NHrevHl68MEHVaNGjWucGQD8H/rQtbFzHwLsiNCJYu3PpzV99dVXatKkiQICAlSqVCl17txZP//8s6Q/TpmaN2+e9u/fLw8PD82dO1eSlJ6erpEjR6pq1ary8fFR7dq1NXv2bJfHuXDhgkaPHq3w8HD5+vqqffv2mj9/vjw8PLRv3z7n/lu1aqXBgwcrJCRE9957r3JycnT8+HENGTJEkZGR8vb2VunSpdWlSxfn/S7OY+DAgZo4caLuvPNO+fn5qUOHDjpy5IjmzJmjatWqKSAgQK1bt3a5X0GuNJ9KlSpp7ty5SklJueLpP2vWrFHTpk0VEBCg8uXLa+DAgTp16pQkKS8vT2PHjlXlypXlcDhUuXJlvfDCC7pw4YKkP5q9JJdPmb/++mudO3dOMTExuvvuu7V69Wrnuj179ujw4cMuzT43N1cffPCBOnbsqE6dOik4OFj/+7//e9n5N2rUSLVr19aCBQvyrUtKStLOnTv1xBNPSJK+//57de3aVWXLlpWXl5fuvPNODRs2TGfPni1w34mJifLw8FBiYqLLeEGn173//vuqVauWHA6HKlasqPHjxysnJ8e5/vjx44qOjla5cuXk4+OjevXqaf78+ZedG4BbD30ov+Leh67GlXpFTEyMWrdurTlz5qh69epyOByqW7euVqxY4bKfzZs3q2nTpvL391fFihU1depUtW7dWjExMZLkPI02Li4u3ym1y5cvV926deXj46Pq1avn60X//Oc/dffdd8vHx0d33nmnnnrqqXxHiFEMGMBG+vXrZyIjI82FCxcKXJo2bWqaNWvm3L5Zs2bO28nJycbX19cMGTLErFu3zixZssTUqFHDVKlSxeTm5ppffvnFdOjQwZQrV85s3rzZHD161GRlZZnatWubsmXLmhkzZphVq1aZQYMGGUlm4sSJzsfp37+/cTgc5tVXXzWrVq0yffv2NQ6Hw0gyv/32m7N2T09P06pVK7N27VqzbNkyk5eXZxo1amSqVatmFi5caNavX28mT55sAgICTJs2bVzmERQUZJo1a2ZWrlxp3n33XePp6Wlq1Khh6tWrZ5YtW2ZmzZpl/P39TYcOHQp9/ooyn507d7o8D6mpqQXua8WKFaZEiRKmc+fO5ssvvzTz58835cqVMy1btjTGGDNp0iRTqlQpM3v2bJOYmGhee+01U7JkSTNu3DjnPqpWrWoGDBjgvD1ixAhTr149Y4wxw4YNM5UrV3aumzlzpilRooQ5fvy4c2z58uVGktm8ebMxxpiBAweaUqVKmaysrEKfA2OMmTx5spFkfvnlF5fxZ5991gQFBZkzZ86YgwcPmqCgINO2bVvz5ZdfmjVr1pjhw4fn+9lLcs5p/fr1RpJZv369y34vfR1efG48PDzMsGHDzOrVq81rr71mfHx8TGxsrHObtm3bOn+2a9euNTExMQXuG8DNRR+iD92IPnRp7yhMUXpFv379THBwsKlZs6ZZtGiRWbFihalfv77x9fU1J0+eNMYY89///tf4+vqaBx980HzxxRdm9uzZJiwszDgcDtOvXz9jjDGbN282ksyAAQOccxk3bpyRZO68804zZ84cs3r1atOsWTPj4eFhdu3aZYwxZtGiRcbb29tMmzbNJCYmmnfffdcEBAQ494vig9AJW+nXr5+RdNmlsGa/aNEiI8n8/vvvzvVbt241L7zwgklPT3fuPzIy0rl+xowZRpL59ttvXeoYMGCA8fHxMSdOnDC//PKL8fDwMG+++abLNu3atcvX7CWZvXv3Orc5cOCAadGihdmwYYPLfYcOHWq8vb1d5uHj4+NsIJfuPzk52Tn29NNPm+Dg4EKfv6LMp6DnoSANGjQw99xzj8vYkiVLzF133WUOHDhg2rVrZ1q3bu2y/p///Kf54IMPnLcHDhxo7r77buftmjVrmueff94Y83+N/OeffzbGGNOnTx/TsGFDl/1169bN5f5bt241kszcuXMvW/uxY8eMt7e3iYuLc47l5OSYChUqmEGDBhljjFm9erVp2rSp87VxUZ06dUzbtm2dt682dKalpRk/Pz/n41z0/vvvG0nmxx9/NMYY43A4zCuvvOJcn5uba0aNGmW++eaby84NgLXoQ/ShS11rH7pS6Cxqr7j4M730Q9Svv/7aSDJLlixx1h0WFmbOnDnj3GbTpk1Gkks4/HNNF0PnypUrnWN79+41kszUqVONMX88f9WrVze5ubnObT788EMzZcqUy84f9sPptbCd8uXLa/v27QUu9957b6H3+8tf/iIfHx81atRII0eO1FdffaV69epp4sSJCgoKKvA+iYmJioyMzPd9kujoaJ07d05btmzR+vXrZYxRjx49XLbp1atXvv35+vqqatWqztsVKlTQunXr9OCDDyolJUXr1q3T9OnTtXHjRp0/f97lvjVr1lSpUqWct8uVK6eyZcuqSpUqzrHQ0FClp6cX+hwUZT5FcfbsWX333Xfq0qWLy3i3bt20Z88eVahQQS1atNBXX32lBx98UG+99ZZ++uknPf300+rTp49z+1atWumnn37SiRMnlJqaqv/+979q166dpD9O5fL29tZXX30l6Y9Tni49penEiRNKSEjQo48+qrS0NKWlpal69eqqUaPGFS/kUKZMGT3yyCMup9iuWbNGBw8edJ5a27ZtW3399dfy9fXVnj179OWXX2rSpEk6evRovp/N1di8ebOysrL0yCOPKCcnx7l06tTJWYcktWjRQuPGjdOjjz6quXPn6tixY3rjjTf0wAMPXPNjA7gx6EN/oA9dex+6kqL2CkkqW7asy880PDxckpwXBVq3bp0efvhhl4sxNWnS5LJXYb7Ugw8+6Pz3xZ91WlqapD961Z49e1S/fn298sor+ve//63HH39czzzzzNVPGrc1Qidsx9vbWw0aNChwCQwMLPR+lSpV0tdff63GjRvrvffeU5s2bRQWFqYXX3xReXl5Bd7n5MmTKleuXL7xi2NpaWnOq6DecccdBW5zqTvuuCPfdyUWLFigihUrKjIyUj169Cj0Kn0F/UJytVfzK8p8irofY0y+OV/q2Wef1fTp05WVlaW//e1vqlmzpurUqaO1a9c6t2nZsqVKlCihTZs2adWqVfL393f+IuLn56cHHnhA69evV0pKilJSUlya/fz583X+/HlNmDBBpUqVci4///yztmzZol27dl12DgMGDNCePXu0fft2SdIHH3ygunXrqn79+pL++C7Q6NGjVbp0adWoUUNPPfWUdu7cKV9f3+v6O2YnTpyQJHXo0EFeXl7OJSwsTJJ08OBBSdLixYs1atQobdu2Tf3791eFChXUvn17/fbbb9f82ABuDPrQ/6EPXXsfupyi9oqLdV6qRIk/fv2/+Jo6duxYgc9TQT+Hgvj7+xe67549e2rhwoUKCAjQ+PHjde+996pKlSpavHhxkfYN+yB0Apdo1KiRPv30U508eVJr165V27ZtNWnSJH3yyScFbl+6dGkdPnw43/ihQ4ck/XHE7OInikePHnXZ5s+3C/Ltt9+qb9++6tq1q1JTU3XixAmtXbtWTZo0udqpFUlR5lMUwcHB8vDwyPdnR7Kzs7Vy5UqdOHFCJUqU0JAhQ/Tdd9/p8OHDmjNnjs6dO6du3bo5Pz0PDQ1VvXr1tHXrVq1Zs0YtWrSQt7e3c39t27bVxo0b9c033yggIMDleZkzZ44aN26s9evXuyzLly9XyZIlr/gpc9u2bRUeHq4FCxYoMzNTn332mfMopyS9+uqreuuttzR16lSlpaUpJSVFS5YsUdmyZQvd58Vf5HJzc13GT58+7fx3SEiIpD9+ySvoKMnQoUOdz/Frr72mffv26aefflJ8fLy+/fZbPfXUU5edF4BbG32IPlQURe0VRREeHl7ga6Eor4+i6NWrl7755hudOHFCH3/8sUJDQxUdHe0SjGF/hE7g/5syZYoqVaqk7OxseXt7q2XLlnrvvfckSampqZKkkiVLutynWbNm2r9/f76/4/Xhhx/Ky8tLjRo10v3336+SJUvq008/ddlm6dKlV6xp06ZNysvLU1xcnPOXhtzcXOdpM4V98n2tijKfoggICFC9evX0+eefu4z/61//UocOHZSamqr77rvPeXrNHXfcoZiYGD399NNKT09XRkaG8z6tWrXSjh079PXXXztPabqoXbt2OnTokD777DO1aNFCXl5ekqQdO3bo+++/V//+/Z1Xhry4dOjQQe3atdOCBQtcwt6flShRQjExMfr000/1xRdfyBij3r17O9d/++23qlWrlmJjYxUcHCxJOnDggH744YdCfy4XjwJcfD1J0qlTp7R7927n7b/85S/y9vbWgQMHXI6OeHt7a/To0frtt9+0f/9+RUREaMmSJZKkGjVq6LnnnlObNm1c9g3g9kIfog8VVVF6RVE1a9ZMK1as0Llz55xj//nPf/Lt4+JRzKvRs2dPde3aVdIfHwT06NFDL7/8snJzcwmdxYynuwsAbhUtW7bU888/ry5duujpp5+Wp6en3n33XTkcDud3JEJCQnTkyBGtXLlS9erVU0xMjN5++2116dJFEyZMUJUqVfTFF19o9uzZGjdunEJCQhQSEqLY2FiNGTNG58+fV926dbVs2TIlJCRIuvyb+MXm+vTTTys2NlanTp3S9OnTnafknDlz5rKnal2tosynqCZMmKBHHnlEPXv2VP/+/XX06FGNGTNGHTt2VL169dSsWTO98cYbCgsL03333acDBw7ozTffVLNmzVw+yW7durWmTp2q8+fP52v2devWVVhYmD7//HO99dZbzvHZs2fLy8tL3bt3L7C2fv36acWKFVqwYIEGDhxY6BxiY2M1ceJETZgwQV27dnX5rlKjRo3097//Xa+++qqaNGmiX375RZMmTVJ2dnahfzz7f/7nfxQREaG4uDgFBwerZMmSio+Pdzk1KTQ0VM8995xefvllZWRkqHnz5jpw4IBefvlleXh4qG7dugoODlZ4eLiGDRumjIwMVa1aVTt27NCKFSs0ZsyYy/9gANyy6EP0oUtt2bJFU6ZMyTfetm1bRUVFXbFXFNULL7ygxYsX66GHHtKoUaOUlpaml156SR4eHi6vjZCQEG3atEkbNmxw+R7n5bRs2VKDBg3S3/72N3Xo0EGnTp3S+PHjddddd11VjbABd17FCLjRrnQ1uz//aYo/3169erW5//77TVBQkPHz8zNNmzY1X3/9tXP9Dz/8YO6++27j5eVl4uPjjTF/XOl0wIABpmzZssbhcJi6deuaWbNmuTzuuXPnzIgRI8wdd9xhfHx8zMMPP2zGjh1rJF3xSnxvv/22qVKlinE4HKZixYqmX79+ZtmyZUaSWb58eYHzKGx/F680dzlFmU9RrhpozB9X9mvYsKFxOBzmzjvvNMOHDzeZmZnGGGMuXLhgxo0bZ6pWrWocDoe54447zBNPPOFyqXlj/rh8vsPhMJUqVSrwMaKjo12uHnj27FkTEhJiHn744ULrOnfunAkJCXFe9v5yWrZsaSSZdevW5dvHkCFDTLly5Yyvr6+pUaOGGTdunImLizMOh8N5BUf96Wp/27ZtM/fdd5/z5zl58mQzcODAfD+/t99+20RFRRlvb28TFhZmevfubfbv3+9cf+jQIRMTE2MqVKhgvL29TdWqVc3EiRNdrhAI4OajD13+uaAP/eFKfUiXufrxnDlznNtdqVcU9Dz99ttv+fazYcMG07hxY+NwOExERIR55513zJ133mmGDh3q3ObNN980ISEhxs/Pz+zfv7/Qn+Wf+960adNMVFSU8fX1NaVLlzaPPvqo2bdvX6HPDezJw5jruOIFgCs6efKkVq5cqfbt2ys0NNQ5/uyzz2r27NnOiwEAAGAF+hAuZ+3atfL29nY5ennq1CmFhYXpjTfe0LBhw9xYHeyC02sBi/n5+WnYsGG65557NHz4cAUEBGjjxo2aNm2aXnjhBXeXBwCwOfoQLmfnzp0aO3as4uPjde+99+r48eN68803FRISUuCf1QGuBUc6gZvgP//5j1566SVt2bJFZ86cUdWqVTVo0CANGTIk36XpAQC40ehDKExeXp4mTZqk+fPnKyUlRQEBAWrevLni4+NVrVo1d5cHmyB0AgAAAAAsw59MAQAAAABYhtAJAAAAALAMoRMAAAAAYBmuXnuLycvL08GDBxUYGMgX+wHgJjLGKDMzUxUqVHD5g+jFHX0JANzHLr2J0HmLOXjwoCIiItxdBgAUW6mpqQoPD3d3GbcM+hIAuN/t3psInbeYwMBASX+8sIKCgtxcDQAUHxkZGYqIiHC+D+MP9CUAcB+79CZC5y3m4qlLQUFBNHcAcANOIXVFXwIA97vde9Pte2IwAAAAAOCWR+gEAAAAAFiG0AkAAAAAsAyhEwAAAABgGUInAAAAAMAyhE4AAAAAgGUInQAAAAAAyxA6AQAAAACWIXQCAAAAACxD6AQAAAAAWIbQCQAAAACwDKETAAAAAGAZT3cXgBuvU6f/+3dCgvvqAADADjot6uRyO6FXwc310u0K2+ZqtgMAu+BIJwAAAADAMoROAAAAAIBlCJ0AAAAAAMsQOgEAAAAAliF0AgAAAAAsQ+gEAAAAAFiG0AkAAAAAsAyhEwAAAABgGUInAAAAAMAyhE4AAAAAgGUInQAAAAAAyxA6AQAAAACWIXQCAAAAACxD6AQAAAAAWIbQCQCAjRw7dkzVqlVTYmJivnWHDh1SWFiY5s6de9PrAgAUX4ROAABsYuPGjWrSpImSk5PzrcvLy1Pv3r11/PhxN1QGACjOCJ0AANjAvHnz9Pjjj2vixIkFrp8wYYLCw8MVERFxkysDABR3hE4AAGygXbt2Sk5OVs+ePfOtW79+vRYvXqwZM2a4oTIAQHHn6e4CAADA9StXrlyB40ePHlX//v21dOlSBQQEXHE/2dnZys7Odt7OyMi4YTUCAIonQicAADZljFGfPn00bNgw1a9fv0j3iY+PV1xcnMWV4aJOizo5/53QK+GK4wBwO+L0WgAAbCo1NVVff/21JkyYoJCQEIWEhCglJUVPPfWUOnbsWOB9xowZo/T0dOeSmpp6k6sGANgNRzoBALCpihUr6ty5cy5jlSpV0vjx4xUTE1PgfRwOhxwOx02oDgBQXHCkEwAAAABgGY50AgBgM8aYQtft27fv5hUCAIA40gkAAAAAsBChswC7du1SmzZtVLp0aZUrV059+/bV8ePHJUlbt25V48aNFRAQoMqVK2vWrFku9503b56qVasmf39/NWjQQJs3b3bHFAAAAADglkDo/JOzZ8/qoYce0n333afDhw8rKSlJJ06cUP/+/XXq1Cl16NBBffv2VVpammbNmqURI0Zo27ZtkqTExEQNHTpU8+bNU1pamnr37q1HHnlEWVlZbp4VAAAAALgHofNPUlJSVLduXY0dO1be3t4KDQ3VwIEDtWHDBi1dulShoaEaMmSIPD091bJlS/Xu3Vtvv/22JOn999/XY489pvvvv19eXl4aMWKEypQpo48++sjNswIAAAAA9yB0/kmNGjW0cuVKlSxZ0jm2ZMkS1a9fX0lJSapTp47L9lFRUdq1a5ckXXF9QbKzs5WRkeGyAAAAAIBdEDovwxijl156SQkJCZo6daoyMzPl7+/vso2fn59Onz4tSVdcX5D4+HgFBwc7l4iIiBs/EQAAAABwE0JnITIyMtS9e3d9+OGH2rBhg+rUqSN/f/9838/MyspSYGCgJF1xfUHGjBmj9PR055KamnrjJwMAAAAAbkLoLEBycrIaNmyojIwM7dixw3nKbO3atZWUlOSy7e7du1W7du0irS+Iw+FQUFCQywIAAAAAdkHo/JNTp06pZcuWuu+++7R69WqVKVPGua5r1646fPiwpkyZogsXLmj9+vVasGCBYmNjJUmxsbFasGCB1q9frwsXLmjKlCk6cuSIunTp4q7pAAAAAIBbETr/ZM6cOUpJSdHHH3+soKAgBQQEOJfQ0FCtWbNGn3zyiUJDQ/XEE09o2rRpatGihSSpVatWmjFjhgYPHqxSpUpp0aJFWrlypUqXLu3mWQEAAACAe3i6u4BbzciRIzVy5MhC1zdo0EAbN24sdH10dLSio6OtKA0AAAAAbjsc6QQAAAAAWIbQCQAAAACwDKETAAAAAGAZvtMJAABgY50WdXK5ndArwU2VACiuONIJAAAAALAMoRMAAAAAYBlCJwAAAADAMoROAAAAAIBlCJ0AAAAAAMsQOgEAAAAAliF0AgAAAAAsQ+gEAAAAAFiG0AkAAAAAsAyhEwAAAABgGUInAAAAAMAyhE4AAAAAgGUInQAAAAAAyxA6AQAAAACWIXQCAGAjx44dU7Vq1ZSYmOgcW7p0qerVq6egoCBVqlRJcXFxysvLc1+RAIBihdAJAIBNbNy4UU2aNFFycrJz7LvvvlOfPn30yiuvKC0tTStXrtTcuXM1efJkN1YKAChOCJ0AANjAvHnz9Pjjj2vixIku4/v27dOgQYPUsWNHlShRQjVr1lSXLl20YcMGN1UKAChuCJ0AANhAu3btlJycrJ49e7qMd+vWTW+99Zbz9tmzZ7V8+XLVr1//ZpcIACimPN1dAAAAuH7lypW74jaZmZnq3r27fH19NWLEiAK3yc7OVnZ2tvN2RkbGDasRAFA8EToBACgGfv75Z3Xr1k1hYWFav369AgMDC9wuPj5ecXFxN7m6W0+nRZ3cXQIA2Aan1wIAYHMrVqxQo0aN1L59e61evVqlSpUqdNsxY8YoPT3duaSmpt7ESgEAdsSRTgAAbGzLli3q0qWL3nnnHcXGxl5xe4fDIYfDcRMqAwAUFxzpBADAxiZNmqQLFy5o2LBhCggIcC4PPfSQu0sDABQTHOkEAMBmjDHOf3/xxRdurAQAAI50AgAAAAAsROgEAAAAAFiG0AkAAAAAsAyhEwAAAABgGUInAAAAAMAyhE4AAAAAgGUInQAAAAAAyxA6AQAAAACWIXQCAAAAACxD6AQAAAAAWIbQCQAAAACwDKETAAAAAGAZQicAAAAAwDKETgAAAACAZTzdXQAAAICddVrUyd0lAIBbcaQTAAAAAGAZQicAAAAAwDKETgAAAACAZQidAAAAAADLEDoBAAAAAJYhdAIAAAAALEPoBAAAAABYhtAJAAAAALAMoRMAAAAAYBlCJwAAAADAMoROAAAAAIBlCJ0AAAAAAMsQOgEAAAAAliF0XsaxY8dUrVo1JSYmOscGDx4sh8OhgIAA5/Lee+8518+bN0/VqlWTv7+/GjRooM2bN7uhcgAAAAC4NRA6C7Fx40Y1adJEycnJLuPbt2/Xe++9p9OnTzuXJ598UpKUmJiooUOHat68eUpLS1Pv3r31yCOPKCsryx1TAAAUQwV9YLp161Y1btxYAQEBqly5smbNmuW+AgEAxQ6hswDz5s3T448/rokTJ7qMZ2dn64cfflCDBg0KvN/777+vxx57TPfff7+8vLw0YsQIlSlTRh999NHNKBsAUMwV9IHpqVOn1KFDB/Xt21dpaWmaNWuWRowYoW3btrmxUgBAcULoLEC7du2UnJysnj17uozv2rVLFy5c0NixYxUWFqbq1avrtddeU15eniQpKSlJderUcblPVFSUdu3aVehjZWdnKyMjw2UBAOBqFfaB6dKlSxUaGqohQ4bI09NTLVu2VO/evfX222+7qVIAQHFD6CxAuXLl5OnpmW88PT1dzZs317Bhw/T777/rww8/1LRp0/Tmm29KkjIzM+Xv7+9yHz8/P50+fbrQx4qPj1dwcLBziYiIuLGTAQAUC4V9YHotH4gCAHAjETqvQps2bbRu3To1a9ZMXl5eatSokYYPH+48fdbf3z/f9zezsrIUGBhY6D7HjBmj9PR055KammrpHAAA9lTYB6ZX+4EoZ+AAAG60/N0Jhfrss8905MgRDRw40DmWnZ0tX19fSVLt2rWVlJTkcp/du3erQ4cOhe7T4XDI4XBYUzAAoNjz9/dXWlqay9jlPhCNj49XXFzcTajs1tBpUSfnvxN6JViyXwAo7jjSeRWMMRoxYoTWrl0rY4w2b96sqVOnOkNobGysFixYoPXr1+vChQuaMmWKjhw5oi5duri5cgBAcVXYB6K1a9cucHvOwAEA3GiEzqvQpUsXTZ48WU899ZQCAgIUHR2tuLg4RUdHS5JatWqlGTNmaPDgwSpVqpQWLVqklStXqnTp0m6uHABQXHXt2lWHDx/WlClTdOHCBa1fv14LFixQbGxsgds7HA4FBQW5LAAAXA9Or70CY4zL7YEDB7qcXvtn0dHRzhAKAIC7hYaGas2aNXrmmWc0duxYlS1bVtOmTVOLFi3cXRoAoJggdAIAYDN//sC0QYMG2rhxo5uqAQAUd5xeCwAAAACwDKETAAAAAGAZQicAAAAAwDKETgAAAACAZQidAAAAAADLEDoBAAAAAJaxXehMTEx0dwkAAFwVehcAwM5sFzq7d++uqlWr6u9//7tSUlLcXQ4AAFdE7wIA2JntQuehQ4f0+uuva8eOHapevbratm2rRYsWKTs7292lAQBQIHoXAMDObBc6vby81K1bN33++edKTU1Vly5d9MYbb6h8+fIaMmSIvv/+e3eXCACAC3oXAMDObBc6Lzp69KgWLlyoefPmaffu3WrRooV8fHzUqlUrTZo0yd3lAQCQD70LAGBHnu4u4EZbvHix5s+frzVr1qhmzZrq37+/vvzyS5UpU0aS9PDDD6tz58564YUX3FwpAAB/oHcBAOzMdqHzqaeeUq9evbR582bVr18/3/rq1atrxIgRbqgMAICC0bsAAHZmu9B56NAh/fbbb7rzzjslSVu2bFFwcLBq1qwpSQoPD1dcXJw7SwQAwAW9CwBgZ7YLnV988YX69OmjjRs3qn79+tq0aZPGjx+vjz76SA899JC7ywMAIB96143XaVEn578TeiUUOH6jH+dW2hcA3EpsdyGh8ePH6/PPP3eenjRy5Eh98sknGj16tJsrAwCgYPQuAICd2S50pqamql27di5jbdu21f79+91UEQAAl0fvAgDYme1CZ2RkpFavXu0ytnbtWkVGRrqpIgAALo/eBQCwM9t9p3PMmDHq3LmzunXrpsjISKWkpOjTTz/VBx984O7SAAAoEL0LAGBntjvS+fjjj2vlypXy9vbWzp075evrqzVr1qhbt27uLg0AgALRuwAAdma7I52S1Lx5czVv3tzdZQAAUGT0LgCAXdkudB48eFCvvPKK9uzZo7y8PJd169atc1NVAAAUjt4FALAz24XOmJgYHTlyRJ06dZKXl5e7ywEA4IroXQAAO7Nd6Ny+fbv27NmjsmXLursUAACKhN4FALAz211IKCQkRD4+Pu4uAwCAIqN3AQDszHah8+WXX1ZMTIy2b9+ulJQUlwUAgFsRvQsAYGe2O732iSeekCQtW7ZMkuTh4SFjjDw8PJSbm+vO0gAAKBC9CwBgZ7YLnb/99pu7SwAA4KrQuwAAdma702sjIyMVGRmpkydP6rvvvlP58uXl6+uryMhId5cGAECB6F0AADuzXeg8evSo7r//fjVu3Fh9+/ZVcnKyqlatqs2bN7u7NAAACmR179q5c6eaNm2qkJAQlS9fXs8884yys7NvyL4BALgS24XO4cOHq06dOkpLS5OXl5dq1qyp0aNH69lnn3V3aQAAFMjK3pWXl6eOHTuqe/fuOnnypLZv367Vq1fr9ddfvwGVAwBwZbb7Tue6dev066+/ys/PTx4eHpKk5557Tm+88YabKwMAoGBW9q5Tp07p0KFDysvLkzFGklSiRAn5+fld974BACgK2x3p9Pb21tmzZyXJ2VwzMzMVGBjozrIAACiUlb0rNDRUI0aM0KhRo+RwOBQREaHq1atrxIgR171vAACKwnah85FHHlF0dLT27t0rDw8PHT16VE899ZQefvhhd5cGAECBrOxdeXl58vX11fTp03XmzBn9+OOP2r17t8aNG1fg9tnZ2crIyHBZAAC4HrYLna+++qoCAgJUo0YNpaWlqXz58srKytKrr77q7tIAACiQlb1r2bJlWrp0qQYPHiyHw6FatWpp3LhxmjFjRoHbx8fHKzg42LlERERcdw23i06LOjkXAMCNY7vvdAYEBOiTTz7RsWPHtG/fPoWHh6t8+fLuLgsAgEJZ2btSUlLyXanWy8tL3t7eBW4/ZswYjRw50nk7IyOjWAVPAMCNZ7vQuWHDBpfbe/fu1d69eyVJTZs2dUdJAABclpW9q127dhozZowmTZqk559/Xvv379crr7yi6OjoArd3OBxyOBzX9ZgAAFzKdqGzefPm+cZKlCihihUr6tdff735BQEAcAVW9q6oqCh9+eWXeumll/T6668rODhY0dHRhX6nEwCAG812oTMvL8/l9vHjxxUXF6dKlSq5pyAAAK7A6t7VunVrtW7d+obsCwCAq2W7Cwn9WZkyZfT6669rypQp7i4FAIAioXcBAOzE9qFT+uMPY587d87dZQAAUGT0LgCAXdju9NrY2FiX2+fPn9c333zDaUUAgFsWvQsAYGe2C53GGJfbPj4+GjZsmAYOHOimigAAuDx6FwDAzmwXOufMmePuEgAAuCr0LgCAndkudE6YMKFI240dO9biSgAAKBp6FwDAzmwXOn/88UctXbpUdevWVY0aNfT7779r48aNuvfeexUYGChJ8vDwoHEDAG4Z9C4AgJ3ZLnSWKFFCkydP1rBhw5xjH374oRISEvTRRx+5sTIAAApG7wIA2Jnt/mTKihUrNGTIEJexXr166auvvnJTRQAAXB69CwBgZ7YLnWXLltWGDRtcxlatWqWIiAg3VQQAwOXRuwAAdma702vHjBmjDh06qHv37qpYsaJ+/fVXJSQk6NNPP3V3aQAAFIjeBQCwM9sd6XziiSeUkJCgkiVL6t///rdCQ0O1efNmtW3b1t2lAQBQIHoXAMDObHekU5Jat26t1q1b6/jx4ypTpoy7ywEA4IroXQAAu7Ldkc4LFy7oxRdfVHBwsCIjI/Xrr7+qYcOGOnTokLtLAwCgQPQuAICd2e5IZ1xcnNatW6dPPvlEPXv2VFhYmMLDw/XMM8/o448/dnd5AADkQ+/C5XRa1KlI2yX0SrDkMa9lv9d7fwD2YrsjnQsWLNCSJUvUtm1beXh4yN/fX3PmzNG6deuuel/Hjh1TtWrVlJiY6BzbunWrGjdurICAAFWuXFmzZs1yuc+8efNUrVo1+fv7q0GDBtq8efP1TgkAYHM3sncBAHCrsV3oPH36tO644w5JkjFGkuTn56cSJa5uqhs3blSTJk2UnJzsHDt16pQ6dOigvn37Ki0tTbNmzdKIESO0bds2SVJiYqKGDh2qefPmKS0tTb1799YjjzyirKysGzQ7AIAd3ajeBQDArch23axJkyaKi4uTJHl4eEiSpk2bpoYNGxZ5H/PmzdPjjz+uiRMnuowvXbpUoaGhGjJkiDw9PdWyZUv17t1bb7/9tiTp/fff12OPPab7779fXl5eGjFihMqUKaOPPvroBs0OAGBHN6J3AQBwq7LddzonT56s1q1ba+7cucrMzFRUVJQyMzP11VdfFXkf7dq1U+/eveXp6anHHnvMOZ6UlKQ6deq4bBsVFeU8xTYpKUmxsbH51u/ates6ZgQAsLsb0bsAALhV2S50litXTklJSVq+fLn27dun8PBwdezYUYGBgVe1j4JkZmbK39/fZczPz0+nT58u0vqCZGdnKzs723k7IyOjyHUCAOzhRvQuAABuVbYLnbVq1dL333+vHj163PB9+/v7Ky0tzWUsKyvL+UuBv79/vu9vZmVlXfbvrcXHxztPqQIAFE9W9i4AANzNdt/plGTZhXtq166tpKQkl7Hdu3erdu3aRVpfkDFjxig9Pd25pKam3vjCAQC3PC46BwCwK9sd6WzRooUaNWqkhx56SBUqVHBekEGSxo4de1377tq1q5577jlNmTJFQ4YM0bfffqsFCxbo888/lyTFxsaqS5cuevTRR/XAAw/o7bff1pEjR9SlS5dC9+lwOORwOK6rLgDA7c3K3gUAgLvZLnTu27dPVatW1Z49e7Rnzx7nuIeHx3U37tDQUK1Zs0bPPPOMxo4dq7Jly2ratGlq0aKFJKlVq1aaMWOGBg8erN9//121atXSypUrVbp06et6XACAvVnZuwAAcDfbhM727dtr1apVWr9+vSTp7Nmz8vX1ve79Xvx7aRc1aNBAGzduLHT76OhoRUdHX/fjAgDsz6reBQDArcQ23+nctGmTy+0777zTTZUAAFA09C4AQHFgm9D5Z38+QgkAwK2O3gUAsCPbhs5LL8IAAMDtgN4FALAj24ZOAAAAAID72eZCQhcuXND8+fOdpyadP3/e5bYk9e3b113lAQCQD70LAFAc2CZ0hoWFuVxWvmzZsi63PTw8aNwAgFvKzepdJ0+e1PDhw7V8+XLl5eWpWbNmeuedd1S+fPnr3jcAAFdim9C5b98+d5cAAMBVuVm9q1u3bipVqpSSk5NVsmRJxcTE6K9//au+/PLLm/L4AIDizTahEwAA5Pfdd99py5YtOnLkiIKCgiRJM2fO1KFDh9xcGQCguOBCQgAA2Ni2bdsUFRWlmTNnqlq1aipfvrxGjRpV6Km12dnZysjIcFkAALgeHOkEAMDGTp48qe+//14NGzbUv//9b2VlZalPnz7q27dvgafXxsfHKy4uzg2VXr9Oizq5u4RbBs8FgFsJRzoBALAxh8MhSZoyZYoCAwMVFhamiRMnasWKFTp9+nS+7ceMGaP09HTnkpqaerNLBgDYDEc6AQCwsaioKOXl5en8+fPy8fGRJOXm5kqSy59mucjhcDiDKgAANwJHOgEAsLE2bdqoSpUqio2N1enTp3Xs2DG9+OKL6ty5swIDA91dHgCgGCB0AgBgY15eXvr666/l6empu+66S9WrV1d4eLhmz57t7tIAAMUEp9cCAGBzFSpU0OLFi91dBgCgmOJIJwAAAADAMoROAAAAAIBlCJ0AAAAAAMsQOgEAAAAAliF0AgAAAAAsQ+gEAAAAAFiG0AkAAAAAsAyhEwAAAABgGUInAAAAAMAyhE4AAAAAgGUInQAAAAAAyxA6AQAAAACWIXQCAAAAACxD6AQAAAAAWMbT3QUAAABYqdOiTu4u4aa43nkW9f6XbpfQK+Gq91uU+wCwF450AgAAAAAsQ+gEAAAAAFiG0AkAAAAAsAyhEwAAAABgGUInAAAAAMAyhE4AAAAAgGUInQAAAAAAyxA6AQAAAACWIXQCAAAAACxD6AQAAAAAWIbQCQAAAACwDKETAAAAAGAZQicAAAAAwDKETgAAAACAZQidAAAUE7m5uWrevLliYmLcXQoAoBghdAIAUEzExcXpm2++cXcZAIBihtAJAEAxsG7dOi1dulTdunVzdykAgGKG0AkAgM0dPXpUAwYM0MKFC+Xn53fZbbOzs5WRkeGyAABwPTzdXQAAALBOXl6eoqOjNXLkSNWtW/eK28fHxysuLu4mVAY76LSok/PfCb0S3FgJgFsZRzoBALCx+Ph4+fj4aOjQoUXafsyYMUpPT3cuqampFlcIALA7jnQCAGBj8+fP18GDBxUSEiJJysrKkiR99tlnSktLy7e9w+GQw+G4iRUCAOyO0AkAgI399NNPLrcv/rmUuXPn3vxiAADFEqfXAgAAAAAsw5FOAACKEY5wAgBuNo50AgAAAAAsQ+gEAAAAAFiG0HkNPvroI3l6eiogIMC59OnTR5K0detWNW7cWAEBAapcubJmzZrl5moBAAAAwH0Inddg+/bt6tOnj06fPu1c5s+fr1OnTqlDhw7q27ev0tLSNGvWLI0YMULbtm1zd8kAAAAA4BaEzmuwfft2NWjQIN/40qVLFRoaqiFDhsjT01MtW7ZU79699fbbb7uhSgAAAABwP0LnVcrLy9POnTu1fPlyRUZGKjw8XE8++aROnTqlpKQk1alTx2X7qKgo7dq1y03VAgAAAIB7ETqv0rFjx3TPPfeoe/fu+u9//6tNmzZp7969io6OVmZmpvz9/V229/Pz0+nTpwvdX3Z2tjIyMlwWAAAAALALQudVCgsL04YNGxQbGys/Pz9VrFhRr7/+ulauXCljjLKysly2z8rKUmBgYKH7i4+PV3BwsHOJiIiwegoAAAAAcNMQOq/S999/r9GjR8sY4xzLzs5WiRIl1KhRIyUlJblsv3v3btWuXbvQ/Y0ZM0bp6enOJTU11bLaAQAAAOBmI3RepdKlS2v69On6xz/+oZycHKWkpOjZZ59VTEyMunfvrsOHD2vKlCm6cOGC1q9frwULFig2NrbQ/TkcDgUFBbksAAAAAGAXhM6rFB4eruXLl+uzzz5T6dKl1aBBAzVs2FDTp09XaGio1qxZo08++UShoaF64oknNG3aNLVo0cLdZQMAAACAW3i6u4DbUbNmzbRp06YC1zVo0EAbN268yRUBAAAAwK2JI50AAAAAAMsQOgEAAAAAluH0WgAAgKvQaVEnd5dwXW7V+i+tK6FXghsrAXCjcaQTAAAAAGAZQicAAAAAwDKETgAAAACAZQidAAAAAADLEDoBAAAAAJYhdAIAAAAALEPoBAAAAABYhtAJAAAAALAMoRMAAAAAYBlCJwAAAADAMoROAAAAAIBlCJ0AAAAAAMsQOgEAAAAAliF0AgAAAAAsQ+gEAMDmdu3apTZt2qh06dIqV66c+vbtq+PHj7u7LABAMUHoBADAxs6ePauHHnpI9913nw4fPqykpCSdOHFC/fv3d3dpAIBigtAJAICNpaSkqG7duho7dqy8vb0VGhqqgQMHasOGDe4uDQBQTHi6uwAAAGCdGjVqaOXKlS5jS5YsUf369QvcPjs7W9nZ2c7bGRkZltYHALA/QicAAMWEMUYvv/yyEhISCj3SGR8fr7i4OMtq6LSok8vthF4Jlj0Wbow//8yA4uzS/w+8fxUdp9cCAFAMZGRkqHv37vrwww+1YcMG1alTp8DtxowZo/T0dOeSmpp6kysFANgNRzoBALC55ORkdejQQRUrVtSOHTtUpkyZQrd1OBxyOBw3sToAgN1xpBMAABs7deqUWrZsqfvuu0+rV6++bOAEAMAKhE4AAGxszpw5SklJ0ccff6ygoCAFBAQ4FwAAbgZCJwAANjZy5EgZY3TmzBmdPn3aZQEA4GbgO5021+mSC84lcIEtAAAAADcZRzoBAAAAAJYhdAIAAAAALEPoBAAAAABYhtAJAAAAALAMoRMAAAAAYBlCJwAAAADAMoROAAAAAIBlCJ0AAAAAAMsQOgEAAAAAliF0AgAAAAAsQ+gEAAAAAFiG0AkAAAAAsAyhEwAAAABgGU93FwAAAOyt06JOlu0roVfCDds3rk9Rf85F2e7SbYr6M76W+xTl/kV9zV3v41vlcvW7u2Z3P75VeJ/KjyOdAAAAAADLEDoBAAAAAJYhdAIAAAAALMN3OouRTpecXp7AqeUAAAAAbgKOdAIAAAAALEPoBAAAAABYhtAJAAAAALAMoRMAAAAAYBlCJwAAAADAMoROAAAAAIBlCJ0AAAAAAMsQOgEAAAAAliF0AgAAAAAsQ+gEAAAAAFiG0GmBo0ePqnPnzgoJCVGZMmU0fPhw5eTkuLusQnXq5LrcyP0BANzvdutLAAB7IXRaoGfPngoICNDBgwe1bds2ffXVV5o8ebK7ywIAFFP0JQCAO3m6uwC7+eWXX5SYmKgDBw7Iz89PVapU0csvv6znnntOzz77rLvLc7rcUchL1yUkXN04AODWcrv0JQCAfRE6b7CkpCSVLl1aFSpUcI5FRUUpJSVFaWlpCgkJcdk+Oztb2dnZztvp6emSpIyMjGuu4cKFa75rPpeWcel+27cv2v0v3e7jj6/+8R99tODxa9kXbg+X/sz5Od8ct/tzfqPqv/i+a4y5zopuLbdEX8oqvDFd7X7/vK9L73+5x8Htq6ivkUt//tfyei3s/pd7zd3Ix7dKUf/PuKNmdz/+tShKzUV9zRSFbXqTwQ01f/58ExER4TL2yy+/GEkmNTU13/bjxo0zklhYWFhYbpGloPfq2xl9iYWFheX2X2733sSRzhvM399fWVlZLmMXbwcGBubbfsyYMRo5cqTzdl5enk6ePKnQ0FB5eHhc1WNnZGQoIiJCqampCgoKuobqb33M0R6Yoz3YbY7GGGVmZrocEbQDd/alm8lur8erxfyZP/O35/zt0psInTdY7dq1deLECR05ckRhYWGSpN27dys8PFzBwcH5tnc4HHI4HC5jfz7V6WoFBQXZ7j/cnzFHe2CO9mCnORb0Pn27uxX60s1kp9fjtWD+zJ/522/+duhNXL32Brvrrrv0wAMPaPjw4crMzNRvv/2mv//97xowYIC7SwMAFEP0JQCAuxE6LbBkyRLl5OSocuXKaty4sdq3b6+XX37Z3WUBAIop+hIAwJ04vdYCYWFh+uSTT2764zocDo0bNy7faVF2whztgTnaQ3GYo124qy/dTMX99cj8mT/zL77zvx14GHO7X38XAAAAAHCr4vRaAAAAAIBlCJ0AAAAAAMsQOgEAAAAAliF02sTRo0fVuXNnhYSEqEyZMho+fLhycnLcXdY1OXbsmKpVq6bExETn2NatW9W4cWMFBASocuXKmjVrlst95s2bp2rVqsnf318NGjTQ5s2bb3LVRbNr1y61adNGpUuXVrly5dS3b18dP35ckn3muG7dOjVu3FhBQUEqV66chg4dqrNnz0qyzxwvys3NVfPmzRUTE+Mcs8scP/roI3l6eiogIMC59OnTR5J95ojb1/W8l9rFtbz/2MXJkyfVt29fhYaGqlSpUurcubMOHTokqXg8Bzt37lTTpk0VEhKi8uXL65lnnlF2drYke8/fzr8fFgsGttC8eXPTu3dvc+bMGZOcnGxq1aplXn/9dXeXddW+/fZbU7VqVSPJrF+/3hhjzMmTJ03p0qXN9OnTzYULF8zatWtNYGCg2bp1qzHGmPXr15vAwEDz7bffmvPnz5u33nrLlClTxpw5c8aNM8kvKyvLlC9f3owdO9ZkZ2eb48ePmw4dOpiOHTvaZo5Hjx41Pj4+Zs6cOSY3N9ccPHjQ1K5d24wdO9Y2c7zUyy+/bEqUKGH69etnjLHPa9UYY0aNGmViYmLyjdtpjrg9Xc97qZ1c7fuPnTRv3tx06dLFnDp1ymRkZJiuXbuahx9+uFg8B7m5uaZ8+fJm6tSpJjc316SmppoaNWqYCRMm2Hr+dv79sLggdNrA3r17jSRz4MAB59jixYtNxYoV3VjV1Zs7d66pWLGiWbx4scubysyZM81dd93lsu2gQYNM3759jTHG9O7d2/z1r391WX/33Xeb2bNn35S6i+qnn34y7du3Nzk5Oc6xzz//3AQFBdlmjsYYk5GRYYwxJi8vz/zwww+mWrVq5p///Ket5miMMWvXrjVRUVGmR48ezl/67DTHpk2bmunTp+cbt9MccXu6nvdSu7iW9x+72LFjh/Hx8THp6enOsRMnTpgff/yxWDwHx48fN5LM5MmTTU5OjklNTTU1a9Y0b7zxhm3nb/ffD4sLTq+1gaSkJJUuXVoVKlRwjkVFRSklJUVpaWnuK+wqtWvXTsnJyerZs6fLeFJSkurUqeMyFhUVpV27dhVp/a2iRo0aWrlypUqWLOkcW7JkierXr2+bOUpSYGCgJCkiIkJ16tRR+fLl1b9/f1vN8ejRoxowYIAWLlwoPz8/57hd5piXl6edO3dq+fLlioyMVHh4uJ588kmdOnXKNnPE7et63kvt4Frff+xi27ZtioqK0syZM1WtWjWVL19eo0aNUvny5YvFcxAaGqoRI0Zo1KhRcjgcioiIUPXq1TVixAjbzt/uvx8WF4ROG8jMzJS/v7/L2MVGdPr0aXeUdE3KlSsnT0/PfOOFze/i3K60/lZkjNFLL72khIQETZ061ZZz3Lt3rw4cOKCSJUuqe/futpljXl6eoqOjNXLkSNWtW9dlnV3meOzYMd1zzz3q3r27/vvf/2rTpk3au3evoqOjbTNH2MPVvpfe7q7n/ccuTp48qe+//1579+7Vv//9b/3nP//RgQMH1Ldv32LxHOTl5cnX11fTp0/XmTNn9OOPP2r37t0aN26cbedfnH4/tDNCpw34+/srKyvLZezi7YtHnW5nhc3v4tyutP5Wk5GRoe7du+vDDz/Uhg0bVKdOHdvNUZJ8fX1VoUIFvfbaa1q1apVt5hgfHy8fHx8NHTo03zq7zDEsLEwbNmxQbGys/Pz8VLFiRb3++utauXKljDG2mCNuf9fyXnq7u573H7twOBySpClTpigwMFBhYWGaOHGiVqxYccX3JztYtmyZli5dqsGDB8vhcKhWrVoaN26cZsyYUWxeAxfZpecWF4ROG6hdu7ZOnDihI0eOOMd2796t8PBwBQcHu7GyG6N27dpKSkpyGdu9e7dq165dpPW3kuTkZDVs2FAZGRnasWOH87QPu8xx06ZNuvvuu3X+/HnnWHZ2try9vRUVFWWLOc6fP1+JiYkKCQlRSEiIFi5cqIULFyokJMQ2P8fvv/9eo0ePljHGOZadna0SJUqoUaNGtpgjbm/X+l56u7ue9x+7iIqKUl5enkufyc3NlSTVq1fP9s9BSkqK80q1F3l5ecnb27vYvAYuskvPLTbc+o1S3DAPPPCAeeyxx0xGRob59ddfTa1atcy4cePcXdY10yVfFD9+/LgJCQkxkydPNufPnzfr1q0zgYGBZt26dcYYY7766ivn7fPnz5vJkyebUqVKmRMnTrhxBvmdPHnSVKxY0cTExJjc3FyXdXaZY2ZmpomIiDAjRoww2dnZZt++faZRo0Zm8ODBtpnjn/Xr1895IQ+7zDE1NdX4+/ub1157zVy4cMHs37/f/OUvfzEDBgywzRxx+7qe91K7uZr3H7s4f/68qVatmunWrZvJzMw0R48eNS1btjRdunQpFs9BUlKScTgcZuLEiSYnJ8ckJyebOnXqmL/97W/FYv52/P2wuCB02sThw4dN9+7dTWhoqClbtqwZNWqUy5X9bjeXvqkYY8z27dvNfffdZwIDA02VKlXMnDlzXLafP3++qVGjhvH39zeNGjUyW7ZsubkFF8Gbb75pJBk/Pz/j7+/vshhjjzka80dDbNOmjQkJCTGRkZHmxRdfNOfOnTPG2GeOl7r0lz5j7DPHxMRE06RJExMYGGjKli1rhg4das6ePWuMsc8ccXu63vdSO7na9x+7OHDggOnZs6cpV66cCQkJMX379jWnTp0yxhSP52DNmjWmcePGJjg42FSsWNG88MILJjs72xhj//nb8ffD4sLDmEvOnwIAAAAA4AbiO50AAAAAAMsQOgEAAAAAliF0AgAAAAAsQ+gEAAAAAFiG0AkAAAAAsAyhEwAAAABgGUInAAAAAMAyhE4AAABAUnp6uo4dO+buMgDbIXQCkCS1bdtWXbt2LXDdzJkzFRYWpuzs7ALX79u3Tx4eHtq3b5+FFQIAigMPDw8lJia65bGrVaumpKQkSdLcuXNVqVIlt9QB2A2hE4AkadiwYUpISNDhw4fzrXvnnXc0aNAgORwON1QGAMDNcfz4cXeXANgSoROAJKlDhw6KjIzU3LlzXca3bNmiH3/8Ue3bt1fHjh1VsWJF+fr6KioqSl9++WWB+/rzp9R//rR4586datGihUqVKqW77rpLkydPljHGglkBAOxm8eLF+p//+R8FBwerfv36+te//uVc17x5c40ZM0ZNmzZVQECAatasqY8//ti5ft++fWrfvr2CgoJ09913a/LkyfLw8JAk1ahRQ5L00EMP6fXXX5ck5eTkaPTo0apYsaKCgoL017/+VTk5OTdxtoA9EDoBSJJKlCihp556Su+//75LAHznnXf06KOPasCAAapTp46Sk5OVnp6udu3aafDgwVf9OAcPHlTLli3VvXt3HT16VJ9//rlmzJih995770ZOBwBgQytWrNCgQYM0ffp0nTx5UnFxcerWrZvzlFhJeu+99zR16lSdPHlS3bp105NPPqlz584pNzdXDz/8sCpUqKBDhw5p9erV+uCDD5z3+/nnnyVJK1eu1HPPPSdJOnDggEJCQvTrr79qy5YtWrhwoT755JObO2nABgidAJwGDBigw4cPa/369ZKkkydP6uOPP9Yzzzyj5cuXa/z48crLy9O+fftUqlQpHThw4Kof48MPP1TNmjU1ZMgQeXl5KSoqSs8++6ymT59+o6cDALCZ6dOna/DgwWratKlKliypjh07qlOnTnr33Xed2/To0UP33HOPvL291a9fP6Wnp+vo0aPasmWL9uzZo3/+85/y9/dXZGSkJk6ceNnHCwoK0vPPPy9PT09FRUWpbt26Sk5OtnqagO14ursAALeO4OBg9enTRzNnzlTLli01e/Zs3XPPPWrYsKGWLVumRx55RIcPH1bNmjVVtmzZazoldt++ffruu+8UEhLiHMvLy1PJkiVv4EwAAHa0b98+JSYm6p133nGO5eTkqFWrVs7b5cqVc/7by8tL0h99JjU1VWXKlJG/v79zfZUqVS77eKVKlXKefitJ3t7enF4LXANCJwAXQ4cO1b333qsTJ07ovffe04QJE3Tw4EH16NFDy5YtU6dOnSRJS5cu1aefflrgPkqUKKHz5887b196YYbw8HC1bNlSq1atco6dOHFCmZmZFs0IAGAX4eHh6tu3r0aPHu0cS0lJka+v7xXvGxkZqWPHjikrK0t+fn6SpP3791tWK4D/w+m1AFxERUXpgQce0MiRI5WVlaVu3bopMzNTubm5zk+Hd+/erQkTJkiSS7i8qGbNmlqyZIlycnKUnJysWbNmOdf17t1bmzdv1oIFC5STk6NDhw6pY8eOGjly5M2ZIADglnfs2DH9/vvvLktOTo6efPJJTZs2Tdu3b5ck7dixQ/Xr19eiRYuuuM/GjRurVq1aGjVqlLKysnTgwAGNHTvWZRuHw6H09HRL5gQUZ4ROAPkMHTpUH3zwgQYPHiwvLy/VqFFD//jHP9S7d28FBwerR48eio2NlZeXl3744Yd8958xY4Z27typ0qVLq2fPnhowYIBzXWRkpFatWqX//d//1R133KG6deuqZs2a+a6aCwAovh599FFFRES4LL/88ou6d++uSZMmqX///goKClL37t01YsQIDR069Ir7LFGihJYsWaI9e/aobNmyatWqlZo1a+Y8BVeSBg4cqF69eunFF1+0cnpAseNh+DsFAAAAsLmzZ89q8+bNatasmfM6AgkJCRo0aNA1XRgPQNFxpBMAAAC25+3trR49emjmzJnKy8vT0aNH9cYbb6hjx47uLg2wPUInAAAAbK9kyZL6/PPPNXfuXJUqVUp16tRR7dq19dZbb7m7NMD2OL0WAAAAAGAZjnQCAAAAACxD6AQAAAAAWIbQCQAAAACwDKETAAAAAGAZQicAAAAAwDKETgAAAACAZQidAAAAAADLEDoBAAAAAJYhdAIAAAAALEPoBAAAAABYhtAJAAAAALAMoRMAAAAAYBlCJwAAAADAMoROAAAAAIBlCJ0AAAAAAMsQOgEAAAAAliF0AgAAAAAsQ+gEAAAAAFiG0AkAAAAAsAyhEwAAAABgGUInAAAAAMAyhE4AAAAAgGUInQAAAAAAyxA6AQAAAACWIXQCAAAAACxD6AQAAAAAWIbQCQAAAACwDKETAAAAAGAZQicAAAAAwDKETgAAAACAZQidAAAAAADLEDoBAAAAAJYhdAIAAAAALEPoBAAAAABYhtAJAAAAALAMoRMAAAAAYBlCJwAAAADAMoROAAAAAIBlCJ0AAAAAAMsQOgEAAAAAliF0AgAAAAAsQ+gEAAAAAFiG0AkAAAAAsAyhEwAAAABgGUInAAAAAMAyhE4AAAAAgGUInQAAAAAAyxA6AQAAAACWIXQCAAAAACzz/wCRpbUJL2ffowAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 1000x400 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot histogram of csWA Values and lengths \n",
    "from matplotlib import pyplot as plt\n",
    "f, ax = plt.subplots(1, 2, figsize=(10, 4))\n",
    "ax[0].hist(csWA['value'], bins=100, color='blue', alpha=0.7)\n",
    "ax[0].set_title('Histogram of csWA Values')\n",
    "ax[0].set_xlabel('Value')\n",
    "ax[0].set_ylabel('Frequency')\n",
    "ax[1].hist(csWA['length'], bins=100, color='green', alpha=0.7)\n",
    "ax[1].set_title('Histogram of csWA Lengths')\n",
    "ax[1].set_xlabel('Length')\n",
    "ax[1].set_ylabel('Frequency')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3ba40728",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'# Export all sense and antisense regions to fasta files \\n\\nsense_csWA_seq_path = os.path.join(csRNAseq_dir, \\'csWA_OR_sense_sequences.fasta\\')\\nantisense_csWA_seq_path = os.path.join(csRNAseq_dir, \\'csWA_OR_antisense_sequences.fasta\\')\\n\\nwith open(sense_csWA_seq_path, \\'w\\') as sense_fasta: \\n    for i, row in csWA[csWA[\\'rel_strand\\'] == \\'sense\\'].iterrows():\\n        sense_fasta.write(f\">{row[\\'location\\']}|{row[\\'gene_name\\']}|{row[\\'strand\\']}\\n\")\\n        sense_fasta.write(f\"{row[\\'sequence\\']}\\n\")\\n\\nwith open(antisense_csWA_seq_path, \\'w\\') as antisense_fasta: \\n    for i, row in csWA[csWA[\\'rel_strand\\'] == \\'antisense\\'].iterrows():\\n        antisense_fasta.write(f\">{row[\\'location\\']}|{row[\\'gene_name\\']}|{row[\\'strand\\']}\\n\")\\n        antisense_fasta.write(f\"{row[\\'sequence\\']}\\n\")'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"# Export all sense and antisense regions to fasta files \n",
    "\n",
    "sense_csWA_seq_path = os.path.join(csRNAseq_dir, 'csWA_OR_sense_sequences.fasta')\n",
    "antisense_csWA_seq_path = os.path.join(csRNAseq_dir, 'csWA_OR_antisense_sequences.fasta')\n",
    "\n",
    "with open(sense_csWA_seq_path, 'w') as sense_fasta: \n",
    "    for i, row in csWA[csWA['rel_strand'] == 'sense'].iterrows():\n",
    "        sense_fasta.write(f\">{row['location']}|{row['gene_name']}|{row['strand']}\\n\")\n",
    "        sense_fasta.write(f\"{row['sequence']}\\n\")\n",
    "\n",
    "with open(antisense_csWA_seq_path, 'w') as antisense_fasta: \n",
    "    for i, row in csWA[csWA['rel_strand'] == 'antisense'].iterrows():\n",
    "        antisense_fasta.write(f\">{row['location']}|{row['gene_name']}|{row['strand']}\\n\")\n",
    "        antisense_fasta.write(f\"{row['sequence']}\\n\")\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "f05ad974",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sense regions: 162\n",
      "Number of antisense regions: 119\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>csRNAscore</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>location</th>\n",
       "      <th>sequence</th>\n",
       "      <th>rel_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr4</td>\n",
       "      <td>10527130</td>\n",
       "      <td>10527230</td>\n",
       "      <td>562.590346</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>Or5-9E108</td>\n",
       "      <td>Chr4:10527130-10527230</td>\n",
       "      <td>CAAGCGCAGTTCCCTTTTCTACGCTATAGTTCGATAGTTCTCGATG...</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr10</td>\n",
       "      <td>1963589</td>\n",
       "      <td>1963689</td>\n",
       "      <td>473.125627</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>Or5-9E241</td>\n",
       "      <td>Chr10:1963589-1963689</td>\n",
       "      <td>CAGTTTATGTGAGCGCAAGCGCAATAGGCAGTTATACTTGCGCGAT...</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr12</td>\n",
       "      <td>8761285</td>\n",
       "      <td>8761385</td>\n",
       "      <td>311.071569</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>Or5-L22</td>\n",
       "      <td>Chr12:8761285-8761385</td>\n",
       "      <td>TCTCGCGCAGACGCCACGGTGACTACGAGGCTCGCTCTTACGTGAT...</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr4</td>\n",
       "      <td>10603156</td>\n",
       "      <td>10603256</td>\n",
       "      <td>101.522030</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>Or5-9E118</td>\n",
       "      <td>Chr4:10603156-10603256</td>\n",
       "      <td>ACGCGCAAGCGCAGTTCCCTTTTCCACTGCCCCAGTAGTTCTTGGC...</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr7</td>\n",
       "      <td>5138088</td>\n",
       "      <td>5138188</td>\n",
       "      <td>95.484292</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>Or5-U64</td>\n",
       "      <td>Chr7:5138088-5138188</td>\n",
       "      <td>CGTTCTACGTACCTTTAACCATCGATATAGTCGGGTCACCCACTTG...</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>Chr11</td>\n",
       "      <td>9784252</td>\n",
       "      <td>9784352</td>\n",
       "      <td>1.050000</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Or5-E21</td>\n",
       "      <td>Chr11:9784252-9784352</td>\n",
       "      <td>TCAACGTAATACGTATCTGACTATATTTGCAATTCAATTATCTTCT...</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>Chr11</td>\n",
       "      <td>13626572</td>\n",
       "      <td>13626672</td>\n",
       "      <td>1.040000</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Or5-9E348</td>\n",
       "      <td>Chr11:13626572-13626672</td>\n",
       "      <td>TATTATATATCGCACATAATAGAATGGTAGTCGGTCGTATTTTAAT...</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>Chr11</td>\n",
       "      <td>13613177</td>\n",
       "      <td>13613277</td>\n",
       "      <td>1.040000</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Or5-9E345</td>\n",
       "      <td>Chr11:13613177-13613277</td>\n",
       "      <td>CATTCTCTGATACCAAGGAATAAAGCCGGACGAGATAAAATTTTGA...</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>Chr11</td>\n",
       "      <td>13606725</td>\n",
       "      <td>13606825</td>\n",
       "      <td>1.040000</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Or5-9E353</td>\n",
       "      <td>Chr11:13606725-13606825</td>\n",
       "      <td>TGCTAAAATATTTATTCATTTCTCATGCCGAAGAATGGAACGAGAA...</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>Chr11</td>\n",
       "      <td>13629281</td>\n",
       "      <td>13629381</td>\n",
       "      <td>1.002060</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Or5-9E349</td>\n",
       "      <td>Chr11:13629281-13629381</td>\n",
       "      <td>AAGCGACGAGACATGCAAAATTGTTTCGCGAGAATTGCCATATGTG...</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>281 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end  csRNAscore strand gene_strand  gene_name  \\\n",
       "0     Chr4  10527130  10527230  562.590346      +           -  Or5-9E108   \n",
       "1    Chr10   1963589   1963689  473.125627      +           -  Or5-9E241   \n",
       "2    Chr12   8761285   8761385  311.071569      +           +    Or5-L22   \n",
       "3     Chr4  10603156  10603256  101.522030      +           -  Or5-9E118   \n",
       "4     Chr7   5138088   5138188   95.484292      -           +    Or5-U64   \n",
       "..     ...       ...       ...         ...    ...         ...        ...   \n",
       "276  Chr11   9784252   9784352    1.050000      -           -    Or5-E21   \n",
       "277  Chr11  13626572  13626672    1.040000      -           -  Or5-9E348   \n",
       "278  Chr11  13613177  13613277    1.040000      -           -  Or5-9E345   \n",
       "279  Chr11  13606725  13606825    1.040000      -           -  Or5-9E353   \n",
       "280  Chr11  13629281  13629381    1.002060      -           -  Or5-9E349   \n",
       "\n",
       "                    location  \\\n",
       "0     Chr4:10527130-10527230   \n",
       "1      Chr10:1963589-1963689   \n",
       "2      Chr12:8761285-8761385   \n",
       "3     Chr4:10603156-10603256   \n",
       "4       Chr7:5138088-5138188   \n",
       "..                       ...   \n",
       "276    Chr11:9784252-9784352   \n",
       "277  Chr11:13626572-13626672   \n",
       "278  Chr11:13613177-13613277   \n",
       "279  Chr11:13606725-13606825   \n",
       "280  Chr11:13629281-13629381   \n",
       "\n",
       "                                              sequence rel_strand  \n",
       "0    CAAGCGCAGTTCCCTTTTCTACGCTATAGTTCGATAGTTCTCGATG...  antisense  \n",
       "1    CAGTTTATGTGAGCGCAAGCGCAATAGGCAGTTATACTTGCGCGAT...  antisense  \n",
       "2    TCTCGCGCAGACGCCACGGTGACTACGAGGCTCGCTCTTACGTGAT...      sense  \n",
       "3    ACGCGCAAGCGCAGTTCCCTTTTCCACTGCCCCAGTAGTTCTTGGC...  antisense  \n",
       "4    CGTTCTACGTACCTTTAACCATCGATATAGTCGGGTCACCCACTTG...  antisense  \n",
       "..                                                 ...        ...  \n",
       "276  TCAACGTAATACGTATCTGACTATATTTGCAATTCAATTATCTTCT...      sense  \n",
       "277  TATTATATATCGCACATAATAGAATGGTAGTCGGTCGTATTTTAAT...      sense  \n",
       "278  CATTCTCTGATACCAAGGAATAAAGCCGGACGAGATAAAATTTTGA...      sense  \n",
       "279  TGCTAAAATATTTATTCATTTCTCATGCCGAAGAATGGAACGAGAA...      sense  \n",
       "280  AAGCGACGAGACATGCAAAATTGTTTCGCGAGAATTGCCATATGTG...      sense  \n",
       "\n",
       "[281 rows x 10 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create another collection with 25bp upstream and downstream of each region \n",
    "\n",
    "distance = 50\n",
    "\n",
    "promoter_regions = [] \n",
    "\n",
    "for i, row in csWA.iterrows():\n",
    "    if row['strand'] == '+':\n",
    "        start = row['start'] - distance\n",
    "        end = row['start'] + distance\n",
    "    else:\n",
    "        start = row['end'] - distance\n",
    "        end = row['end'] + distance\n",
    "    promoter_regions.append({\n",
    "        'chrom': row['chrom'],\n",
    "        'start': start,\n",
    "        'end': end,\n",
    "        'csRNAscore': row['value'],\n",
    "        'strand': row['strand'],\n",
    "        'gene_strand': row['gene_strand'],\n",
    "        'gene_name': row['gene_name'],\n",
    "        'location': f\"{row['chrom']}:{start}-{end}\"\n",
    "    })\n",
    "\n",
    "promoter_regions = pd.DataFrame(promoter_regions)\n",
    "promoter_regions['sequence'] = promoter_regions.apply(\n",
    "    lambda row: get_sequence(genome_seq, row['chrom'], row['start'], row['end'], row['strand']), axis=1\n",
    ")\n",
    "promoter_regions['rel_strand'] = np.where(promoter_regions['strand'] == promoter_regions['gene_strand'], 'sense', 'antisense')\n",
    "\n",
    "# how many sense and antisense regions are there?\n",
    "print(f\"Number of sense regions: {np.sum(promoter_regions['rel_strand'] == 'sense')}\")\n",
    "print(f\"Number of antisense regions: {np.sum(promoter_regions['rel_strand'] == 'antisense')}\")\n",
    "\n",
    "promoter_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "74cba8df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>csRNAscore</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>location</th>\n",
       "      <th>sequence</th>\n",
       "      <th>rel_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr12</td>\n",
       "      <td>8761285</td>\n",
       "      <td>8761385</td>\n",
       "      <td>311.071569</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>Or5-L22</td>\n",
       "      <td>Chr12:8761285-8761385</td>\n",
       "      <td>TCTCGCGCAGACGCCACGGTGACTACGAGGCTCGCTCTTACGTGAT...</td>\n",
       "      <td>sense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>Chr12</td>\n",
       "      <td>8761154</td>\n",
       "      <td>8761254</td>\n",
       "      <td>2.480000</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>Or5-L22</td>\n",
       "      <td>Chr12:8761154-8761254</td>\n",
       "      <td>AATCGATTGACTAAAGCTATCGGACTTGTTACTTGGGTAACAATGT...</td>\n",
       "      <td>antisense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom    start      end  csRNAscore strand gene_strand gene_name  \\\n",
       "2    Chr12  8761285  8761385  311.071569      +           +   Or5-L22   \n",
       "139  Chr12  8761154  8761254    2.480000      -           +   Or5-L22   \n",
       "\n",
       "                  location                                           sequence  \\\n",
       "2    Chr12:8761285-8761385  TCTCGCGCAGACGCCACGGTGACTACGAGGCTCGCTCTTACGTGAT...   \n",
       "139  Chr12:8761154-8761254  AATCGATTGACTAAAGCTATCGGACTTGTTACTTGGGTAACAATGT...   \n",
       "\n",
       "    rel_strand  \n",
       "2        sense  \n",
       "139  antisense  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get Or5-L22\n",
    "promoter_regions[promoter_regions['gene_name'] == 'Or5-L22']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "38a8ce8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export all sense and antisense regions to fasta files \n",
    "\n",
    "sense_promoter_seq_path = os.path.join(csRNAseq_dir, 'csWA_OR_promoter_sense_sequences.fasta')\n",
    "antisense_promoter_seq_path = os.path.join(csRNAseq_dir, 'csWA_OR_promoter_antisense_sequences.fasta')\n",
    "\n",
    "with open(sense_promoter_seq_path, 'w') as sense_fasta: \n",
    "    for i, row in promoter_regions[promoter_regions['rel_strand'] == 'sense'].iterrows():\n",
    "        sense_fasta.write(f\">{row['location']}|{row['gene_name']}|{row['strand']}\\n\")\n",
    "        sense_fasta.write(f\"{row['sequence']}\\n\")\n",
    "\n",
    "with open(antisense_promoter_seq_path, 'w') as antisense_fasta: \n",
    "    for i, row in promoter_regions[promoter_regions['rel_strand'] == 'antisense'].iterrows():\n",
    "        antisense_fasta.write(f\">{row['location']}|{row['gene_name']}|{row['strand']}\\n\")\n",
    "        antisense_fasta.write(f\"{row['sequence']}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "d583a19c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['meme', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_sense_sequences.fasta', '-oc', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_sense_sequences_meme', '-dna', '-nmotifs', '10', '-evt', '0.01', '-minw', '10', '-maxw', '20', '-minsites', '10', '-maxiter', '1000', '-mod', 'zoops']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The output directory '../raw-data/OBir/csRNAseq/csWA_OR_promoter_sense_sequences_meme' already exists.\n",
      "Its contents will be overwritten.\n",
      "BACKGROUND: using background model of order 0\n",
      "PRIMARY (classic): n 162 p0 162 p1 0 p2 0\n",
      "SEQUENCE GROUP USAGE-- Starts/EM: p0; Trim: p0; pvalue: p0; nsites: p0,p1,p2\n",
      "SEEDS: maxwords 16200 highwater mark: seq 162 pos 90\n",
      "BALANCE: samples 162 chars 16200 nodes 1 chars/node 16200\n",
      "Initializing the motif probability tables for 2 to 162 sites...\n",
      "nsites = 162\n",
      "Done initializing.\n",
      "\n",
      "seqs=   162, min_w= 100, max_w=  100, total_size=    16200\n",
      "\n",
      "motif=1\n",
      "SEED DEPTHS: 10 20 40 80 160 162\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 162, iter= 130 \n",
      "motif=2\n",
      "SEED DEPTHS: 10 20 40 80 160 162\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 162, iter= 180 \n",
      "motif=3\n",
      "SEED DEPTHS: 10 20 40 80 160 162\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 162, iter= 160 \n",
      "motif=4\n",
      "SEED DEPTHS: 10 20 40 80 160 162\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 162, iter= 230 \n",
      "motif=5\n",
      "SEED DEPTHS: 10 20 40 80 160 162\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 162, iter= 950 \n",
      "motif=6\n",
      "SEED DEPTHS: 10 20 40 80 160 162\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 162, iter= 350 \n",
      "motif=7\n",
      "SEED DEPTHS: 10 20 40 80 160 162\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 162, iter= 290 \n",
      "motif=8\n",
      "SEED DEPTHS: 10 20 40 80 160 162\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 162, iter= 110 \n",
      "motif=9\n",
      "SEED DEPTHS: 10 20 40 80 160 162\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 162, iter= 100 \n",
      "motif=10\n",
      "SEED DEPTHS: 10 20 40 80 160 162\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 162, iter= 170 \n",
      "motif=11\n",
      "SEED DEPTHS: 10 20 40 80 160 162\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 162, iter= 440 \n",
      "motif=12\n",
      "SEED DEPTHS: 10 20 40 80 160 162\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 162, iter= 610 \n",
      "motif=13\n",
      "SEED DEPTHS: 10 20 40 80 160 162\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 162, iter= 250 \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['meme', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_sense_sequences.fasta', '-oc', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_sense_sequences_meme', '-dna', '-nmotifs', '10', '-evt', '0.01', '-minw', '10', '-maxw', '20', '-minsites', '10', '-maxiter', '1000', '-mod', 'zoops'], returncode=0)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# run MEME on sense sequences\n",
    "\n",
    "import subprocess\n",
    "\n",
    "input_path = os.path.join(csRNAseq_dir, 'csWA_OR_promoter_sense_sequences.fasta')\n",
    "output_dir = os.path.join(csRNAseq_dir, 'csWA_OR_promoter_sense_sequences_meme')\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    " # MEME command\n",
    "meme_command = [\n",
    "    \"meme\", input_path, \n",
    "    \"-oc\", output_dir, \n",
    "    \"-dna\", \n",
    "    \"-nmotifs\", \"10\", # number of motifs # \"-maxsize\", \"1000000\", # increases the maximum size MEME will consider \n",
    "    \"-evt\", \"0.01\", # E-value threshold \n",
    "    \"-minw\", \"10\", # minimum motif width \n",
    "    \"-maxw\", \"20\", # maximum motif width\n",
    "    \"-minsites\", \"10\", # minimum number of sites required for a motif \n",
    "    \"-maxiter\", \"1000\", # maximum number of iterations \n",
    "    \"-mod\", \"zoops\", # options: zoops (0 or 1), oops (1), anr (any) \n",
    "]\n",
    "\n",
    "print(meme_command) \n",
    "subprocess.run(meme_command, check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "95faed49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['meme', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_antisense_sequences.fasta', '-oc', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_antisense_sequences_meme', '-dna', '-nmotifs', '10', '-evt', '0.01', '-minw', '10', '-maxw', '20', '-minsites', '10', '-maxiter', '1000', '-mod', 'zoops']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The output directory '../raw-data/OBir/csRNAseq/csWA_OR_promoter_antisense_sequences_meme' already exists.\n",
      "Its contents will be overwritten.\n",
      "BACKGROUND: using background model of order 0\n",
      "PRIMARY (classic): n 119 p0 119 p1 0 p2 0\n",
      "SEQUENCE GROUP USAGE-- Starts/EM: p0; Trim: p0; pvalue: p0; nsites: p0,p1,p2\n",
      "SEEDS: maxwords 11900 highwater mark: seq 119 pos 90\n",
      "BALANCE: samples 119 chars 11900 nodes 1 chars/node 11900\n",
      "Initializing the motif probability tables for 2 to 119 sites...\n",
      "nsites = 119\n",
      "Done initializing.\n",
      "\n",
      "seqs=   119, min_w= 100, max_w=  100, total_size=    11900\n",
      "\n",
      "motif=1\n",
      "SEED DEPTHS: 10 20 40 80 119\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 119, iter=  30 \n",
      "motif=2\n",
      "SEED DEPTHS: 10 20 40 80 119\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 119, iter= 130 \n",
      "motif=3\n",
      "SEED DEPTHS: 10 20 40 80 119\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 119, iter= 120 \n",
      "motif=4\n",
      "SEED DEPTHS: 10 20 40 80 119\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 119, iter= 100 \n",
      "motif=5\n",
      "SEED DEPTHS: 10 20 40 80 119\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 119, iter= 140 \n",
      "motif=6\n",
      "SEED DEPTHS: 10 20 40 80 119\n",
      "SEED WIDTHS: 10 14 19 20\n",
      "em: w=  20, psites= 119, iter= 200 \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['meme', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_antisense_sequences.fasta', '-oc', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_antisense_sequences_meme', '-dna', '-nmotifs', '10', '-evt', '0.01', '-minw', '10', '-maxw', '20', '-minsites', '10', '-maxiter', '1000', '-mod', 'zoops'], returncode=0)"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "input_path = os.path.join(csRNAseq_dir, 'csWA_OR_promoter_antisense_sequences.fasta')\n",
    "output_dir = os.path.join(csRNAseq_dir, 'csWA_OR_promoter_antisense_sequences_meme')\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    " # MEME command\n",
    "meme_command = [\n",
    "    \"meme\", input_path, \n",
    "    \"-oc\", output_dir, \n",
    "    \"-dna\", \n",
    "    \"-nmotifs\", \"10\", # number of motifs # \"-maxsize\", \"1000000\", # increases the maximum size MEME will consider \n",
    "    \"-evt\", \"0.01\", # E-value threshold \n",
    "    \"-minw\", \"10\", # minimum motif width \n",
    "    \"-maxw\", \"20\", # maximum motif width\n",
    "    \"-minsites\", \"10\", # minimum number of sites required for a motif \n",
    "    \"-maxiter\", \"1000\", # maximum number of iterations \n",
    "    \"-mod\", \"zoops\", # options: zoops (0 or 1), oops (1), anr (any) \n",
    "]\n",
    "\n",
    "print(meme_command) \n",
    "subprocess.run(meme_command, check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "4889118c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_pspm(pspm):\n",
    "    \"\"\"\n",
    "    Normalize a Position-Specific Probability Matrix (PSPM) so that each column sums to 1.\n",
    "    If a column sums to 0.9, add 0.1 to the smallest value.\n",
    "    If a column sums to 1.1, subtract 0.1 from the largest value.\n",
    "    \n",
    "    Parameters:\n",
    "        pspm (numpy.ndarray): A 4 x n matrix representing the PSPM.\n",
    "    \n",
    "    Returns:\n",
    "        numpy.ndarray: The normalized PSPM.\n",
    "    \"\"\"\n",
    "    # Ensure the input is a numpy array\n",
    "    pspm = np.array(pspm)\n",
    "    \n",
    "    # Iterate over each column in the PSPM\n",
    "    for col_idx in range(pspm.shape[1]):\n",
    "        column = pspm[:, col_idx]\n",
    "        col_sum = np.sum(column)\n",
    "        \n",
    "        if np.isclose(col_sum, 0.9, atol=1e-3):  # Check if column sum is close to 0.9\n",
    "            min_idx = np.argmin(column)  # Index of the smallest value in the column\n",
    "            pspm[min_idx, col_idx] += 0.1  # Add 0.1 to the smallest value\n",
    "            \n",
    "        elif np.isclose(col_sum, 1.1, atol=1e-3):  # Check if column sum is close to 1.1\n",
    "            max_idx = np.argmax(column)  # Index of the largest value in the column\n",
    "            pspm[max_idx, col_idx] -= 0.1  # Subtract 0.1 from the largest value\n",
    "            \n",
    "        # If column sums to neither 0.9 nor 1.1, leave it unchanged\n",
    "    \n",
    "    return pspm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "ad51e13a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_meme_output(file_path):\n",
    "    with open(file_path, 'r') as file:\n",
    "        content = file.read()\n",
    "    \n",
    "    # Define regex patterns to extract relevant information\n",
    "    motif_pattern = re.compile(r\"MOTIF\\s+(\\S+)\\s+MEME-\\d+\\s+width\\s=\\s+(\\d+)\\s+sites\\s=\\s+(\\d+)\\s+llr\\s=\\s+(\\d+)\\s+E-value\\s=\\s+([\\d\\.\\-e]+)\")\n",
    "    \n",
    "    # Regex patterns to extract the lines corresponding to each base (A, C, G, T)\n",
    "    line_A_pattern = re.compile(r\"Simplified\\s+A\\s+([:a-zA-Z0-9]+)\")\n",
    "    line_C_pattern = re.compile(r\"pos.-specific\\s+C\\s+([:a-zA-Z0-9]+)\")\n",
    "    line_G_pattern = re.compile(r\"probability\\s+G\\s+([:a-zA-Z0-9]+)\")\n",
    "    line_T_pattern = re.compile(r\"matrix\\s+T\\s+([:a-zA-Z0-9]+)\")\n",
    "\n",
    "    # Regex pattern to extract numbers from each line\n",
    "    number_pattern = re.compile(r\"(\\d+)\")\n",
    "\n",
    "    # Find all motifs that match the motif pattern\n",
    "    motifs = motif_pattern.findall(content)\n",
    "    \n",
    "    # Create a DataFrame to store results\n",
    "    motif_data = pd.DataFrame(motifs, columns=['Motif', 'Width', 'Sites', 'LLR', 'E-value'])\n",
    "    \n",
    "    # Convert numeric columns to appropriate types\n",
    "    motif_data['Width'] = pd.to_numeric(motif_data['Width'])\n",
    "    motif_data['Sites'] = pd.to_numeric(motif_data['Sites'])\n",
    "    motif_data['LLR'] = pd.to_numeric(motif_data['LLR'])\n",
    "    motif_data['E-value'] = pd.to_numeric(motif_data['E-value'])\n",
    "\n",
    "    # Initialize a list to store the extracted PSPMs\n",
    "    pspsm_data = []\n",
    "\n",
    "    # Function to transform a string representation to a list of floats based on the rules\n",
    "    def transform_pspm_line(line):\n",
    "        transformed = []\n",
    "        for char in line:\n",
    "            if char == ':':\n",
    "                transformed.append(0.0)\n",
    "            elif char == 'a':\n",
    "                transformed.append(1.0)\n",
    "            elif char.isdigit():\n",
    "                transformed.append(int(char) / 10.0)  # Convert number to a decimal\n",
    "        return transformed\n",
    "\n",
    "    # Extract PSPMs for each motif\n",
    "    motif_blocks = re.split(r'MOTIF\\s+\\S+\\s+MEME-\\d+', content)[1:]  # Split content by motif occurrences\n",
    "    for block in motif_blocks:\n",
    "        # Extract each row corresponding to A, C, G, and T\n",
    "        row_A = line_A_pattern.search(block)\n",
    "        row_C = line_C_pattern.search(block)\n",
    "        row_G = line_G_pattern.search(block)\n",
    "        row_T = line_T_pattern.search(block)\n",
    "        \n",
    "        if row_A and row_C and row_G and row_T:\n",
    "            # Transform each line using the custom rules\n",
    "            A_values = transform_pspm_line(row_A.group(1))\n",
    "            C_values = transform_pspm_line(row_C.group(1))\n",
    "            G_values = transform_pspm_line(row_G.group(1))\n",
    "            T_values = transform_pspm_line(row_T.group(1))\n",
    "\n",
    "            # Combine into a matrix\n",
    "            pspsm_data.append(np.array([A_values, C_values, G_values, T_values]))\n",
    "        else:\n",
    "            print(\"Error extracting PSPM data.\")\n",
    "            pspsm_data.append(None)  # In case extraction fails for some motif\n",
    "\n",
    "    motif_data['PSPM_ACGT'] = pspsm_data # pspsm_data #normalize_pspm(pspsm_data)\n",
    "\n",
    "    return motif_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "8d93864a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Motif</th>\n",
       "      <th>Width</th>\n",
       "      <th>Sites</th>\n",
       "      <th>LLR</th>\n",
       "      <th>E-value</th>\n",
       "      <th>PSPM_ACGT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RCGCAAGCGCARTTSVYWTT</td>\n",
       "      <td>20</td>\n",
       "      <td>65</td>\n",
       "      <td>1169</td>\n",
       "      <td>3.400000e-298</td>\n",
       "      <td>[[0.3, 0.0, 0.0, 0.0, 1.0, 0.9, 0.0, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAGTCBAKTBYGYGCADTCG</td>\n",
       "      <td>20</td>\n",
       "      <td>63</td>\n",
       "      <td>918</td>\n",
       "      <td>9.200000e-192</td>\n",
       "      <td>[[0.0, 1.0, 0.0, 0.0, 0.2, 0.0, 0.7, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTWCRCATSTDYSTTAAAMR</td>\n",
       "      <td>20</td>\n",
       "      <td>46</td>\n",
       "      <td>623</td>\n",
       "      <td>1.500000e-99</td>\n",
       "      <td>[[0.1, 0.0, 0.5, 0.0, 0.7, 0.0, 0.8, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RMKCWDTASKTCTTSGCGTT</td>\n",
       "      <td>20</td>\n",
       "      <td>26</td>\n",
       "      <td>434</td>\n",
       "      <td>9.700000e-71</td>\n",
       "      <td>[[0.5, 0.3, 0.0, 0.0, 0.6, 0.3, 0.0, 1.0, 0.0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATACSSTTSKACKC</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>172</td>\n",
       "      <td>1.900000e-05</td>\n",
       "      <td>[[1.0, 0.1, 1.0, 0.0, 0.2, 0.0, 0.0, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Motif  Width  Sites   LLR        E-value  \\\n",
       "0  RCGCAAGCGCARTTSVYWTT     20     65  1169  3.400000e-298   \n",
       "1  CAGTCBAKTBYGYGCADTCG     20     63   918  9.200000e-192   \n",
       "2  GTWCRCATSTDYSTTAAAMR     20     46   623   1.500000e-99   \n",
       "3  RMKCWDTASKTCTTSGCGTT     20     26   434   9.700000e-71   \n",
       "4        ATACSSTTSKACKC     14     14   172   1.900000e-05   \n",
       "\n",
       "                                           PSPM_ACGT  \n",
       "0  [[0.3, 0.0, 0.0, 0.0, 1.0, 0.9, 0.0, 0.0, 0.0,...  \n",
       "1  [[0.0, 1.0, 0.0, 0.0, 0.2, 0.0, 0.7, 0.0, 0.0,...  \n",
       "2  [[0.1, 0.0, 0.5, 0.0, 0.7, 0.0, 0.8, 0.0, 0.0,...  \n",
       "3  [[0.5, 0.3, 0.0, 0.0, 0.6, 0.3, 0.0, 1.0, 0.0,...  \n",
       "4  [[1.0, 0.1, 1.0, 0.0, 0.2, 0.0, 0.0, 0.0, 0.0,...  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_motifs = pd.DataFrame(columns=['Motif', 'Width', 'Sites', 'LLR', 'E-value', 'PSPM_ACGT'])\n",
    "output_path = os.path.join(csRNAseq_dir, 'csWA_OR_promoter_antisense_sequences_meme', 'meme.txt')\n",
    "motif_data = parse_meme_output(output_path)    \n",
    "motif_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "b6185371",
   "metadata": {},
   "outputs": [],
   "source": [
    "# normalize PSPM \n",
    "motif_data['PSPM_ACGT'] = motif_data['PSPM_ACGT'].apply(normalize_pspm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "124aacf6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1.])"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motif_data['PSPM_ACGT'].iloc[0].sum(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "fc1039bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Motif</th>\n",
       "      <th>Width</th>\n",
       "      <th>Sites</th>\n",
       "      <th>LLR</th>\n",
       "      <th>E-value</th>\n",
       "      <th>PSPM_ACGT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RCGCAAGCGCARTTSVYWTT</td>\n",
       "      <td>20</td>\n",
       "      <td>65</td>\n",
       "      <td>1169</td>\n",
       "      <td>3.400000e-298</td>\n",
       "      <td>[[0.3, 0.0, 0.0, 0.0, 1.0, 0.8, 0.0, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAGTCBAKTBYGYGCADTCG</td>\n",
       "      <td>20</td>\n",
       "      <td>63</td>\n",
       "      <td>918</td>\n",
       "      <td>9.200000e-192</td>\n",
       "      <td>[[0.0, 1.0, 0.0, 0.0, 0.2, 0.1, 0.7, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTWCRCATSTDYSTTAAAMR</td>\n",
       "      <td>20</td>\n",
       "      <td>46</td>\n",
       "      <td>623</td>\n",
       "      <td>1.500000e-99</td>\n",
       "      <td>[[0.1, 0.0, 0.5, 0.0, 0.7, 0.0, 0.8, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RMKCWDTASKTCTTSGCGTT</td>\n",
       "      <td>20</td>\n",
       "      <td>26</td>\n",
       "      <td>434</td>\n",
       "      <td>9.700000e-71</td>\n",
       "      <td>[[0.5, 0.3, 0.0, 0.0, 0.6, 0.3, 0.1, 1.0, 0.0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATACSSTTSKACKC</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>172</td>\n",
       "      <td>1.900000e-05</td>\n",
       "      <td>[[1.0, 0.1, 1.0, 0.0, 0.2, 0.0, 0.0, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Motif  Width  Sites   LLR        E-value  \\\n",
       "0  RCGCAAGCGCARTTSVYWTT     20     65  1169  3.400000e-298   \n",
       "1  CAGTCBAKTBYGYGCADTCG     20     63   918  9.200000e-192   \n",
       "2  GTWCRCATSTDYSTTAAAMR     20     46   623   1.500000e-99   \n",
       "3  RMKCWDTASKTCTTSGCGTT     20     26   434   9.700000e-71   \n",
       "4        ATACSSTTSKACKC     14     14   172   1.900000e-05   \n",
       "\n",
       "                                           PSPM_ACGT  \n",
       "0  [[0.3, 0.0, 0.0, 0.0, 1.0, 0.8, 0.0, 0.0, 0.0,...  \n",
       "1  [[0.0, 1.0, 0.0, 0.0, 0.2, 0.1, 0.7, 0.0, 0.0,...  \n",
       "2  [[0.1, 0.0, 0.5, 0.0, 0.7, 0.0, 0.8, 0.0, 0.0,...  \n",
       "3  [[0.5, 0.3, 0.0, 0.0, 0.6, 0.3, 0.1, 1.0, 0.0,...  \n",
       "4  [[1.0, 0.1, 1.0, 0.0, 0.2, 0.0, 0.0, 0.0, 0.0,...  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motif_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "68f95d90",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop all rows except the first \n",
    "#motif_data = motif_data.iloc[[0]].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f457bde2",
   "metadata": {},
   "source": [
    "## Write meme file for FIMO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "4fe870cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the output directory and file path\n",
    "output_dir = os.path.join(csRNAseq_dir, 'csWA_OR_promoter_antisense_sequences_meme')\n",
    "output_file = os.path.join(output_dir, 'motifs.meme')\n",
    "\n",
    "# Ensure the output directory exists\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Open the output file for writing in MEME format\n",
    "with open(output_file, 'w') as f:\n",
    "    # Write the header for the MEME file\n",
    "    f.write(\"MEME version 5.0\\n\\n\")\n",
    "    f.write(\"ALPHABET= ACGT\\n\\n\")\n",
    "    f.write(\"strands: + -\\n\\n\")\n",
    "    f.write(\"Background letter frequencies (uniform):\\n\")\n",
    "    f.write(\"A 0.25 C 0.25 G 0.25 T 0.25\\n\\n\")\n",
    "    \n",
    "    # Iterate through the motif data and write each motif\n",
    "    for i, row in motif_data.iterrows():\n",
    "        # Write the MOTIF header\n",
    "        f.write(f\"MOTIF {row['Motif']}\\n\")\n",
    "        f.write(f\"letter-probability matrix: alength=4 w={row[\"Width\"]} nsites={row['Sites']} E={row['E-value']}\\n\")\n",
    "        \n",
    "        # Write the PSPM matrix\n",
    "        np.savetxt(f, row['PSPM_ACGT'].T, fmt=\"%.1f\", delimiter=\"\\t\")\n",
    "        f.write(\"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "343a9b97",
   "metadata": {},
   "source": [
    "Export region upstream of all ORs to a fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "37f04c65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Minimum OR sequence length: 1000\n",
      "Maximum OR sequence length: 1000\n"
     ]
    }
   ],
   "source": [
    "# Get sequences for each OR region of interest\n",
    "OR_rois['seq'] = OR_rois.apply(lambda row: get_sequence(genome_seq, row['chrom'], row['start'], row['end'], row['strand']), axis=1)\n",
    "\n",
    "# Verify length of sequences \n",
    "or_lengths = OR_rois['seq'].apply(len)\n",
    "print(f\"Minimum OR sequence length: {or_lengths.min()}\")\n",
    "print(f\"Maximum OR sequence length: {or_lengths.max()}\")\n",
    "\n",
    "# Export to Fasta file \n",
    "or_fasta_path = os.path.join(csRNAseq_dir, 'ORs_1000bp_upstream_CDS.fasta')\n",
    "with open(or_fasta_path, 'w') as or_fasta:\n",
    "    for i, row in OR_rois.iterrows():\n",
    "        or_fasta.write(f\">{row['gene_name']}|{row['chrom']}:{row['start']}-{row['end']}|{row['strand']}\\n\")\n",
    "        or_fasta.write(f\"{row['seq']}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "399cbabe",
   "metadata": {},
   "source": [
    "### FIMO\n",
    "FIMO scans a set of sequences for individual matches to each of the motifs you provide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "912e2974",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['fimo', '--oc', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_antisense_sequences_fimo', '--thresh', '0.0001', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_antisense_sequences_meme/motifs.meme', '../raw-data/OBir/csRNAseq/ORs_1000bp_upstream_CDS.fasta']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using motif +RCGCAAGCGCARTTSVYWTT of width 20.\n",
      "Using motif -RCGCAAGCGCARTTSVYWTT of width 20.\n",
      "Finding best site passing the output threshold in each of the 498 sequences.\n",
      "Found a best site in 250 sequences.\n",
      "Computing q-values.\n",
      "# Computing q-values.\n",
      "#   Estimating pi_0 from a uniformly sampled set of 10000 p-values.\n",
      "#   Estimating pi_0.\n",
      "# Minimal pi_zero = 1.0003\n",
      "#   Estimated pi_0=1\n",
      "Using motif +CAGTCBAKTBYGYGCADTCG of width 20.\n",
      "Using motif -CAGTCBAKTBYGYGCADTCG of width 20.\n",
      "Finding best site passing the output threshold in each of the 498 sequences.\n",
      "Found a best site in 215 sequences.\n",
      "Computing q-values.\n",
      "# Computing q-values.\n",
      "#   Estimating pi_0 from a uniformly sampled set of 10000 p-values.\n",
      "#   Estimating pi_0.\n",
      "# Minimal pi_zero = 1.00061\n",
      "#   Estimated pi_0=1\n",
      "Using motif +GTWCRCATSTDYSTTAAAMR of width 20.\n",
      "Using motif -GTWCRCATSTDYSTTAAAMR of width 20.\n",
      "Finding best site passing the output threshold in each of the 498 sequences.\n",
      "Found a best site in 222 sequences.\n",
      "Computing q-values.\n",
      "# Computing q-values.\n",
      "#   Estimating pi_0 from a uniformly sampled set of 10000 p-values.\n",
      "#   Estimating pi_0.\n",
      "# Minimal pi_zero = 0.8836\n",
      "#   Estimated pi_0=0.887706\n",
      "Using motif +RMKCWDTASKTCTTSGCGTT of width 20.\n",
      "Using motif -RMKCWDTASKTCTTSGCGTT of width 20.\n",
      "Finding best site passing the output threshold in each of the 498 sequences.\n",
      "Found a best site in 100 sequences.\n",
      "Computing q-values.\n",
      "# Computing q-values.\n",
      "#   Estimating pi_0 from a uniformly sampled set of 10000 p-values.\n",
      "#   Estimating pi_0.\n",
      "# Minimal pi_zero = 1.0009\n",
      "#   Estimated pi_0=1\n",
      "Using motif +ATACSSTTSKACKC of width 14.\n",
      "Using motif -ATACSSTTSKACKC of width 14.\n",
      "Finding best site passing the output threshold in each of the 498 sequences.\n",
      "Found a best site in 88 sequences.\n",
      "Computing q-values.\n",
      "# Computing q-values.\n",
      "#   Estimating pi_0 from a uniformly sampled set of 10000 p-values.\n",
      "#   Estimating pi_0.\n",
      "# Minimal pi_zero = 0.998985\n",
      "#   Estimated pi_0=0.999497\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['fimo', '--oc', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_antisense_sequences_fimo', '--thresh', '0.0001', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_antisense_sequences_meme/motifs.meme', '../raw-data/OBir/csRNAseq/ORs_1000bp_upstream_CDS.fasta'], returncode=0)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_dir_path = os.path.join(csRNAseq_dir, 'csWA_OR_promoter_antisense_sequences_fimo')\n",
    "sequence_path = os.path.join(csRNAseq_dir, 'ORs_1000bp_upstream_CDS.fasta')\n",
    "motif_path = os.path.join(csRNAseq_dir, 'csWA_OR_promoter_antisense_sequences_meme', 'motifs.meme')\n",
    "\n",
    "# Ensure the output directory exists\n",
    "os.makedirs(output_dir_path, exist_ok=True)\n",
    "\n",
    "# Run the FIMO command\n",
    "fimo_command = [\n",
    "    \"fimo\",\n",
    "    \"--oc\", output_dir_path,\n",
    "    \"--thresh\", \"0.0001\", # Set the threshold for motif matches\n",
    "    motif_path,\n",
    "    sequence_path\n",
    "]\n",
    "\n",
    "print(fimo_command)\n",
    "subprocess.run(fimo_command, check=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b199189c",
   "metadata": {},
   "source": [
    "## Read in FIMO TSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "8301a019",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of motifs on + strand: 242\n",
      "Number of motifs on - strand: 776\n",
      "Number of motifs with FDR q-value < 0.05: 508 out of 1018 total motifs.\n",
      "Number of motifs with q-value < 0.1 and strand is -: 532 out of 1018 total motifs.\n",
      "Number of unique gene_name with q-value < 0.1 and strand is -: 224 out of 393 total unique gene_name.\n",
      "Number of unique 9E genes with a hit with q-value < 0.1 and strand is -: 209 out of 224.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif_id</th>\n",
       "      <th>motif_alt_id</th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>p-value</th>\n",
       "      <th>q-value</th>\n",
       "      <th>matched_sequence</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>nine_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RCGCAAGCGCARTTSVYWTT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Or5-9E56|Chr2</td>\n",
       "      <td>16707481.0</td>\n",
       "      <td>16707500.0</td>\n",
       "      <td>-</td>\n",
       "      <td>31.32930</td>\n",
       "      <td>8.250000e-13</td>\n",
       "      <td>8.190000e-08</td>\n",
       "      <td>GCGCAAGCGCAATTGACTTT</td>\n",
       "      <td>Or5-9E56</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RCGCAAGCGCARTTSVYWTT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Or5-9E57|Chr2</td>\n",
       "      <td>16713011.0</td>\n",
       "      <td>16713030.0</td>\n",
       "      <td>-</td>\n",
       "      <td>31.32930</td>\n",
       "      <td>8.250000e-13</td>\n",
       "      <td>8.190000e-08</td>\n",
       "      <td>GCGCAAGCGCAATTGACTTT</td>\n",
       "      <td>Or5-9E57</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RCGCAAGCGCARTTSVYWTT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Or5-9E58|Chr2</td>\n",
       "      <td>16720178.0</td>\n",
       "      <td>16720197.0</td>\n",
       "      <td>-</td>\n",
       "      <td>31.32930</td>\n",
       "      <td>8.250000e-13</td>\n",
       "      <td>8.190000e-08</td>\n",
       "      <td>GCGCAAGCGCAATTGACTTT</td>\n",
       "      <td>Or5-9E58</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RCGCAAGCGCARTTSVYWTT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Or5-9E59|Chr2</td>\n",
       "      <td>16725445.0</td>\n",
       "      <td>16725464.0</td>\n",
       "      <td>-</td>\n",
       "      <td>31.32930</td>\n",
       "      <td>8.250000e-13</td>\n",
       "      <td>8.190000e-08</td>\n",
       "      <td>GCGCAAGCGCAATTGACTTT</td>\n",
       "      <td>Or5-9E59</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RCGCAAGCGCARTTSVYWTT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Or5-9E20|Chr2</td>\n",
       "      <td>2463443.0</td>\n",
       "      <td>2463462.0</td>\n",
       "      <td>-</td>\n",
       "      <td>31.03660</td>\n",
       "      <td>3.190000e-12</td>\n",
       "      <td>8.190000e-08</td>\n",
       "      <td>GCGCAAGCGCAATTGATTTT</td>\n",
       "      <td>Or5-9E20</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1013</th>\n",
       "      <td>ATACSSTTSKACKC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Or5-V4|Chr3</td>\n",
       "      <td>14760355.0</td>\n",
       "      <td>14760368.0</td>\n",
       "      <td>-</td>\n",
       "      <td>7.43000</td>\n",
       "      <td>9.940000e-05</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>ACACCCTTGTGCAC</td>\n",
       "      <td>Or5-V4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1014</th>\n",
       "      <td>RCGCAAGCGCARTTSVYWTT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Or5-9E281|Chr10</td>\n",
       "      <td>12837625.0</td>\n",
       "      <td>12837644.0</td>\n",
       "      <td>-</td>\n",
       "      <td>-8.32927</td>\n",
       "      <td>9.960000e-05</td>\n",
       "      <td>3.450000e-01</td>\n",
       "      <td>TCCCAAGCGTAAGTGAAATT</td>\n",
       "      <td>Or5-9E281</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1015</th>\n",
       "      <td>RCGCAAGCGCARTTSVYWTT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Or5-9E283|Chr10</td>\n",
       "      <td>12846824.0</td>\n",
       "      <td>12846843.0</td>\n",
       "      <td>-</td>\n",
       "      <td>-8.32927</td>\n",
       "      <td>9.960000e-05</td>\n",
       "      <td>3.450000e-01</td>\n",
       "      <td>TCCCAAGCGTAAGTGAAATT</td>\n",
       "      <td>Or5-9E283</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1016</th>\n",
       "      <td>GTWCRCATSTDYSTTAAAMR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Or5-U67|Chr8</td>\n",
       "      <td>1272067.0</td>\n",
       "      <td>1272086.0</td>\n",
       "      <td>-</td>\n",
       "      <td>3.44444</td>\n",
       "      <td>1.000000e-04</td>\n",
       "      <td>3.070000e-01</td>\n",
       "      <td>ATTCGCTACTTTATTGAAGG</td>\n",
       "      <td>Or5-U67</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1017</th>\n",
       "      <td>GTWCRCATSTDYSTTAAAMR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Or5-9E270|Chr10</td>\n",
       "      <td>12768379.0</td>\n",
       "      <td>12768398.0</td>\n",
       "      <td>-</td>\n",
       "      <td>3.44444</td>\n",
       "      <td>1.000000e-04</td>\n",
       "      <td>3.070000e-01</td>\n",
       "      <td>GTTTATAAGTTTATAAAATA</td>\n",
       "      <td>Or5-9E270</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1018 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  motif_id  motif_alt_id    sequence_name       start  \\\n",
       "0     RCGCAAGCGCARTTSVYWTT           NaN    Or5-9E56|Chr2  16707481.0   \n",
       "1     RCGCAAGCGCARTTSVYWTT           NaN    Or5-9E57|Chr2  16713011.0   \n",
       "2     RCGCAAGCGCARTTSVYWTT           NaN    Or5-9E58|Chr2  16720178.0   \n",
       "3     RCGCAAGCGCARTTSVYWTT           NaN    Or5-9E59|Chr2  16725445.0   \n",
       "4     RCGCAAGCGCARTTSVYWTT           NaN    Or5-9E20|Chr2   2463443.0   \n",
       "...                    ...           ...              ...         ...   \n",
       "1013        ATACSSTTSKACKC           NaN      Or5-V4|Chr3  14760355.0   \n",
       "1014  RCGCAAGCGCARTTSVYWTT           NaN  Or5-9E281|Chr10  12837625.0   \n",
       "1015  RCGCAAGCGCARTTSVYWTT           NaN  Or5-9E283|Chr10  12846824.0   \n",
       "1016  GTWCRCATSTDYSTTAAAMR           NaN     Or5-U67|Chr8   1272067.0   \n",
       "1017  GTWCRCATSTDYSTTAAAMR           NaN  Or5-9E270|Chr10  12768379.0   \n",
       "\n",
       "            stop strand     score       p-value       q-value  \\\n",
       "0     16707500.0      -  31.32930  8.250000e-13  8.190000e-08   \n",
       "1     16713030.0      -  31.32930  8.250000e-13  8.190000e-08   \n",
       "2     16720197.0      -  31.32930  8.250000e-13  8.190000e-08   \n",
       "3     16725464.0      -  31.32930  8.250000e-13  8.190000e-08   \n",
       "4      2463462.0      -  31.03660  3.190000e-12  8.190000e-08   \n",
       "...          ...    ...       ...           ...           ...   \n",
       "1013  14760368.0      -   7.43000  9.940000e-05  1.000000e+00   \n",
       "1014  12837644.0      -  -8.32927  9.960000e-05  3.450000e-01   \n",
       "1015  12846843.0      -  -8.32927  9.960000e-05  3.450000e-01   \n",
       "1016   1272086.0      -   3.44444  1.000000e-04  3.070000e-01   \n",
       "1017  12768398.0      -   3.44444  1.000000e-04  3.070000e-01   \n",
       "\n",
       "          matched_sequence  gene_name  nine_exon  \n",
       "0     GCGCAAGCGCAATTGACTTT   Or5-9E56       True  \n",
       "1     GCGCAAGCGCAATTGACTTT   Or5-9E57       True  \n",
       "2     GCGCAAGCGCAATTGACTTT   Or5-9E58       True  \n",
       "3     GCGCAAGCGCAATTGACTTT   Or5-9E59       True  \n",
       "4     GCGCAAGCGCAATTGATTTT   Or5-9E20       True  \n",
       "...                    ...        ...        ...  \n",
       "1013        ACACCCTTGTGCAC     Or5-V4      False  \n",
       "1014  TCCCAAGCGTAAGTGAAATT  Or5-9E281       True  \n",
       "1015  TCCCAAGCGTAAGTGAAATT  Or5-9E283       True  \n",
       "1016  ATTCGCTACTTTATTGAAGG    Or5-U67      False  \n",
       "1017  GTTTATAAGTTTATAAAATA  Or5-9E270       True  \n",
       "\n",
       "[1018 rows x 12 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fimo_results_path = os.path.join(output_dir_path, 'fimo.tsv')\n",
    "fimo_results = pd.read_csv(fimo_results_path, sep='\\t')\n",
    "# Drop rows with nan sequence_name \n",
    "fimo_results = fimo_results.dropna(subset=['sequence_name']).reset_index(drop=True)\n",
    "# Get gene_name \n",
    "fimo_results['sequence_name'] = fimo_results['sequence_name'].astype(str)  # Ensure sequence_name is string type\n",
    "fimo_results['gene_name'] = fimo_results['sequence_name'].apply(lambda x: x.split('|')[0])\n",
    "# Print strand results \n",
    "print(f\"Number of motifs on + strand: {fimo_results[fimo_results['strand'] == '+'].shape[0]}\")\n",
    "print(f\"Number of motifs on - strand: {fimo_results[fimo_results['strand'] == '-'].shape[0]}\")\n",
    "\n",
    "# Print number of hits with FDR q-value < 0.05 \n",
    "fimo_results['q-value'] = fimo_results['q-value'].astype(float)  # Ensure q-value is float type\n",
    "print(f\"Number of motifs with FDR q-value < 0.05: {fimo_results[fimo_results['q-value'] < 0.05].shape[0]} out of {fimo_results.shape[0]} total motifs.\")\n",
    "\n",
    "# Print number of hits with q-value < 0.1 and strand is - \n",
    "print(f\"Number of motifs with q-value < 0.1 and strand is -: {fimo_results[(fimo_results['q-value'] < 0.1) & (fimo_results['strand'] == '-')].shape[0]} out of {fimo_results.shape[0]} total motifs.\")\n",
    "\n",
    "# Print number of unique gene_name with q-value < 0.1 and strand is -\n",
    "print(f\"Number of unique gene_name with q-value < 0.1 and strand is -: {fimo_results[(fimo_results['q-value'] < 0.1) & (fimo_results['strand'] == '-')]['gene_name'].nunique()} out of {fimo_results['gene_name'].nunique()} total unique gene_name.\")\n",
    "\n",
    "# Print number of unique 9E genes with a hit with q-value < 0.1 and strand is -\n",
    "fimo_results['nine_exon'] = fimo_results['gene_name'].apply(lambda x: '9E' in x)\n",
    "print(f\"Number of unique 9E genes with a hit with q-value < 0.1 and strand is -: {fimo_results[(fimo_results['q-value'] < 0.1) & (fimo_results['strand'] == '-') & (fimo_results['nine_exon'])]['gene_name'].nunique()} out of {fimo_results[(fimo_results['q-value'] < 0.1) & (fimo_results['strand'] == '-')]['gene_name'].nunique()}.\")\n",
    "\n",
    "fimo_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28393f92",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e69b7aa7",
   "metadata": {},
   "source": [
    "## SEA\n",
    "Simple Enrichment Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "7c28e8ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['sea', '--oc', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_antisense_sequences_meme/sea', '--m', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_antisense_sequences_meme/motifs.meme', '--p', '../raw-data/OBir/csRNAseq/ORs_500bp_upstream_CDS.fasta']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "# Checking alphabets in 1 motif files.\n",
      "# Loading motifs from file '../raw-data/OBir/csRNAseq/csWA_OR_promoter_antisense_sequences_meme/motifs.meme'\n",
      "# Alphabet: DNA\n",
      "# Positive sequences \"../raw-data/OBir/csRNAseq/ORs_500bp_upstream_CDS.fasta\" - training: 449 hold-out: 49\n",
      "# Negative sequences are shuffled primary sequences (2-order) - training: 449 hold-out: 49\n",
      "# Estimating background model from control sequences.\n",
      "# Background: A 0.341 C 0.159 G 0.159 T 0.341\n",
      "# Background order: 2 Background size: 84\n",
      "# Using Fisher Exact test for p-values.\n",
      "# Computing q-values.\n",
      "#   Cannot estimate pi_0 accurately from fewer than 100 p-values.\n",
      "#   Total p-values = 5. Using pi_zero = 1.0.\n",
      "# Positive sequences \"../raw-data/OBir/csRNAseq/ORs_500bp_upstream_CDS.fasta\" - training: 449 hold-out: 49\n",
      "# Negative sequences are shuffled primary sequences (2-order) - training: 449 hold-out: 49\n",
      "# Estimating background model from control sequences.\n",
      "# Background: A 0.341 C 0.159 G 0.159 T 0.341\n",
      "# Background order: 2 Background size: 84\n",
      "# Using Fisher Exact test for p-values.\n",
      "# Freeing storage...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['sea', '--oc', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_antisense_sequences_meme/sea', '--m', '../raw-data/OBir/csRNAseq/csWA_OR_promoter_antisense_sequences_meme/motifs.meme', '--p', '../raw-data/OBir/csRNAseq/ORs_500bp_upstream_CDS.fasta'], returncode=0)"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_dir_path = os.path.join(csRNAseq_dir, 'csWA_OR_promoter_antisense_sequences_meme', 'sea')\n",
    "sequence_path = os.path.join(csRNAseq_dir, 'ORs_500bp_upstream_CDS.fasta')\n",
    "motif_path = os.path.join(csRNAseq_dir, 'csWA_OR_promoter_antisense_sequences_meme', 'motifs.meme')\n",
    "\n",
    "# Ensure the output directory exists\n",
    "os.makedirs(output_dir_path, exist_ok=True)\n",
    "\n",
    "# Run the SEA command\n",
    "sea_command = [\n",
    "    \"sea\",\n",
    "    \"--oc\", output_dir_path,\n",
    "    \"--m\", motif_path,\n",
    "    \"--p\", sequence_path\n",
    "]\n",
    "\n",
    "print(sea_command)\n",
    "subprocess.run(sea_command, check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cae17350",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "antman",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
